In [2]:
import numpy as np
import pandas as pd
import re
from tqdm import tqdm

# import tensorflow as tf

import requests
import json


In [3]:
# PATH = 'D:/project/데이터 분석 프로젝트/이노포스트/data'
PATH = 'G:/다른 컴퓨터/My_desktop/project/데이터 분석 프로젝트/이노포스트/data'

df = pd.read_csv(PATH + '/03_금연구역/금연구역지정현황(원본).csv', encoding ='cp949')

In [4]:
df.head()

구  구분      시설명                          주소
0  중구  학교   굿모닝유치원  대전광역시 중구 어덕마을로 84-12 (중촌동)
1  중구  학교    금동유치원  대전광역시 중구 태평로152번길 63 (태평동)
2  중구  학교   꿈내리유치원       대전광역시 중구 당디로 45 (산성동)
3  중구  학교  남대전고등학교  대전광역시 중구 보문로111번길 63 (대사동)
4  중구  학교   대전고등학교      대전광역시 중구 대흥로 110 (대흥동)

## data check

In [5]:
df

구    구분      시설명                          주소
0       중구    학교   굿모닝유치원  대전광역시 중구 어덕마을로 84-12 (중촌동)
1       중구    학교    금동유치원  대전광역시 중구 태평로152번길 63 (태평동)
2       중구    학교   꿈내리유치원       대전광역시 중구 당디로 45 (산성동)
3       중구    학교  남대전고등학교  대전광역시 중구 보문로111번길 63 (대사동)
4       중구    학교   대전고등학교      대전광역시 중구 대흥로 110 (대흥동)
...    ...   ...      ...                         ...
45602  유성구  복합용도       식당          대전광역시 유성구 학하동676-1
45603  유성구  복합용도      NaN         대전광역시 유성구 학하동713-12
45604  유성구  복합용도      NaN          대전광역시 유성구 학하동716-7
45605  유성구  복합용도      NaN          대전광역시 유성구 화암동157-6
45606  유성구  복합용도      NaN          대전광역시 유성구 화암동63-10

[45607 rows x 4 columns]

In [6]:
df.isnull().sum()

구         0
구분        0
시설명    1737
주소        1
dtype: int64

In [7]:
df[df['주소'].isnull()] # 황금알식당은 폐업한 상점으로 제거

구               구분    시설명   주소
3417  중구  음식점(휴게, 일반, 제과)  황금알식당  NaN

In [8]:
df = df.dropna(subset=['주소'])
df.isnull().sum()

구         0
구분        0
시설명    1737
주소        0
dtype: int64

In [8]:
# 시설명이 없는 행 추출
# df_null = df[df['시설명'].isnull()]
# df_null

구                 구분  시설명                   주소
12555   서구  연면적 1천제곱미터 이상 건축물  NaN   대전광역시 서구 복수동 258-0
12556   서구  연면적 1천제곱미터 이상 건축물  NaN   대전광역시 서구 복수동 271-1
12558   서구  연면적 1천제곱미터 이상 건축물  NaN  대전광역시 서구 복수동 274-13
12559   서구  연면적 1천제곱미터 이상 건축물  NaN  대전광역시 서구 복수동 275-18
12560   서구  연면적 1천제곱미터 이상 건축물  NaN   대전광역시 서구 복수동 276-0
...    ...                ...  ...                  ...
45601  유성구               복합용도  NaN    대전광역시 유성구 하기동88-2
45603  유성구               복합용도  NaN  대전광역시 유성구 학하동713-12
45604  유성구               복합용도  NaN   대전광역시 유성구 학하동716-7
45605  유성구               복합용도  NaN   대전광역시 유성구 화암동157-6
45606  유성구               복합용도  NaN   대전광역시 유성구 화암동63-10

[1737 rows x 4 columns]

시설명이 없는 건물을 삭제 후 진행했으나 추후 시설명 결측치를 제거하는게 옳은지 고민되어 이 부분을 다시 진행함

In [9]:
# df2 = df[~df['시설명'].isnull()]
# df2.reset_index(drop=True, inplace=True)
# df2.isnull().sum()

구      0
구분     0
시설명    0
주소     0
dtype: int64

### 주소지

In [9]:
df.groupby('구분')['주소'].count().sort_values(ascending=False) 
# 금연구역에도 목욕장이 지정되어 있음, 목욕장 데이터와 중복

구분
음식점                  12556
일반음식점 및 휴게음식점         7718
음식점(휴게, 일반, 제과)       4247
학원                    2173
연면적 1천제곱미터 이상 건축물     2105
의료기관                  1906
담배판매업소 및 담배자동판매기      1367
복합건축물                 1206
어린이집                  1206
어린이놀이시설               1173
담배소매업소                1082
사회복지시설                 991
어린이운송용차량               771
체육시설                   727
담배소매인업소                702
신고체육시설                 680
어린이운송용 승합자동차           673
복합용도                   584
게임제공업소                 559
학교교과교습학원               437
학교                     424
공장                     339
사무용건축물                 277
실내체육시설                 251
청사                     234
체육시설업                  205
어린이운송용승합차              165
게임산업                   136
목욕장                    119
유치원                    104
식품자동판매기영업소              98
게임제공업소시설                94
어린이운송용 및 교통시설           82
도서관                     42
대규모점포                   32
지역아동센터                  24
공연장                     2

### `()`와 `()안의 문자` 제거

In [11]:
df3 = df.copy()

In [12]:
# 번지주소 컬럼 생성
df3.insert(4, '번지주소', df3['주소'])

In [13]:
# ()와 (내용물) 제거
df3["주소"] = df3["주소"].str.replace(pat=r'\([^)]*\)', repl=r' ', regex=True)

In [14]:
# ()가 남아있는지 확인
df3[df3['주소'].str.contains('\)')]

구            구분              시설명  \
260     중구      학교교과교습학원         목양해법영어학원   
739     중구  어린이운송용 승합자동차     무궁화태권도장(태평관)   
877     중구        게임제공업소               쿵스   
1226    중구        실내체육시설           스타복싱클럽   
1947    중구       담배소매인업소              홈마트   
...    ...           ...              ...   
41819  유성구            학원         노은파워리딩학원   
42026  유성구            학원  플레이팩토앤(N)창의교육학원   
42027  유성구            학원          재테크경매학원   
42057  유성구            학원        워릭프랭클린어학원   
42923  유성구          어린이집        KNF새빛어린이집   

                                             주소  \
260    대전광역시 중구 목동로 37 , 312호의 1/4, 313~316호  )   
739                    대전광역시 중구 계백로1615번길 50  )   
877                대전광역시 중구 중앙로156번길 16  앤비백화점)   
1226                    대전광역시 중구 목중로 58, 3층중촌동)   
1947                대전광역시 중구 대둔산로 390   케이티 1층)   
...                                         ...   
41819           대전광역시 유성구 노은로 144, 404호 405호  )   
42026     대전광역시 유성구 지족로 374 , 301-2호, 302호  상가)   
42027      대전광역시 유성구 지족로 374 , 410호,411호일부  상가)   
42057       대전광역시 유성구 하기로 34-29 , 1층일부, 2층전체  )   
42923                대전광역시 유성구 대덕대로989번길 242  )   

                                                    번지주소  
260    대전광역시 중구 목동로 37 , 312호의 1/4, 313~316호 (목동, 목양마...  
739            대전광역시 중구 계백로1615번길 50 (유천동, 무궁화태권도장(태평관))  
877                대전광역시 중구 중앙로156번길 16 (은행동, (3층)앤비백화점)  
1226                             대전광역시 중구 목중로 58, 3층중촌동)  
1947                  대전광역시 중구 대둔산로 390  (산성동.(주)케이티 1층)  
...                                                  ...  
41819       대전광역시 유성구 노은로 144, 404호 405호 (지족동,노은타운(씨네위))  
42026  대전광역시 유성구 지족로 374 , 301-2호, 302호 (지족동, 옥타브(B동)상가)  
42027    대전광역시 유성구 지족로 374 , 410호,411호일부 (지족동,옥타브(B동)상가)  
42057  대전광역시 유성구 하기로 34-29 , 1층일부, 2층전체 (하기동, 하기동 88-...  
42923          대전광역시 유성구 대덕대로989번길 242 (덕진동, 한전원자력연료(주))  

[942 rows x 5 columns]

In [15]:
# 남아있는 ) 제거
df3["주소"] = df3["주소"].str.replace(pat=r'\)', repl=r' ', regex=True)

In [16]:
df3[df3['주소'].str.contains('\)')]

Empty DataFrame
Columns: [구, 구분, 시설명, 주소, 번지주소]
Index: []

In [ ]:
# @ 제거
df3['kakao 주소'] = df3['kakao 주소'].str.replace('@', '아파트')
df3[df3['주소'].str.contains('@')]

### 중복된 공백 제거

In [17]:
print(df3['주소'].str.contains('  ').sum()) # 스페이스바 두번
print(df3['주소'].str.contains('   ').sum()) # 스페이스바 세번
print(df3['주소'].str.contains('    ').sum()) # 스페이스바 네번
print(df3['주소'].str.contains('      ').sum()) # 스페이스바 다섯번
print(df3['주소'].str.contains('       ').sum()) # 스페이스바 여섯번
print(df3['주소'].str.contains('.').sum())
print(df3['주소'].str.contains(',').sum())
print(df3['주소'].str.contains('외국').sum())
print(df3['주소'].str.contains('번지').sum()) # 번지 들어가면 번지주소?
print(df3['주소'].str.contains('동').sum()) # 동 들어가면 번지주소?

23492
2567
21
2
0
45606
12361
12
7950
30658


In [18]:
df3["주소"] = df3["주소"].str.replace(pat='  ', repl=r' ', regex=True)
print(df3['주소'].str.contains('  ').sum()) # 스페이스바 두번
print(df3['주소'].str.contains('   ').sum()) # 스페이스바 세번
print(df3['주소'].str.contains('    ').sum()) # 스페이스바 네번
print(df3['주소'].str.contains('     ').sum()) # 스페이스바 다섯번

2567
8
0
0


In [19]:
df3["주소"] = df3["주소"].str.replace(pat='  ', repl=r' ', regex=True)
print(df3['주소'].str.contains('  ').sum()) # 스페이스바 두번
print(df3['주소'].str.contains('   ').sum()) # 스페이스바 세번

8
0


In [20]:
df3["주소"] = df3["주소"].str.replace(pat='  ', repl=r' ', regex=True)
print(df3['주소'].str.contains('  ').sum()) # 스페이스바 두번
print(df3['주소'].str.contains('   ').sum()) # 스페이스바 세번

0
0


이후 중복된 공백은 여러번 실행으로 0이 될때까지 진행

In [21]:
df_space = df3
df_space[df_space['주소'].str.contains('  ')].count()

구       0
구분      0
시설명     0
주소      0
번지주소    0
dtype: int64

### 주소 형태 체크

In [22]:
df_space['주소'].nunique()

39270

In [23]:
df_space['주소']

0        대전광역시 중구 어덕마을로 84-12 
1        대전광역시 중구 태평로152번길 63 
2             대전광역시 중구 당디로 45 
3        대전광역시 중구 보문로111번길 63 
4            대전광역시 중구 대흥로 110 
                 ...          
45602       대전광역시 유성구 학하동676-1
45603      대전광역시 유성구 학하동713-12
45604       대전광역시 유성구 학하동716-7
45605       대전광역시 유성구 화암동157-6
45606       대전광역시 유성구 화암동63-10
Name: 주소, Length: 45606, dtype: object

### 시 분리

In [24]:
# 시작이 전부 대전광역시인지 확인
df_space['주소'].str[:5].unique()

array(['대전광역시', '대전 중구', '대전시 중', '대저너시 ', '대전시 충', '중구 서문', '대전 서구',
       '서울시 마', '보문산로3', '선화로43', '대전천서로', '부용로41', '당디로 9', '단재로 4',
       '유천로18', '충무로15', '대종로 2', '사득로 2', '천근로 3', '대종로 4', '대종로 1',
       '당골길 2', '테미로 5', '범골로 9', '성산로 2', '성산로14', '오류로 1', '충무로 1',
       '보문산로 ', '대종로55', '계백로15', '봉소루로 ', '돌다리로 ', '산서로 1', '계백로16',
       '대흥로 1', '문창로 2', '봉소루로5', '태평로14', '테미로 2', '계백로 1', '천근로 4',
       '대둔산로 ', '문화로10', '중앙로 2', '중구 오류', '중구 계백', '중구 계룡', '중구 중앙',
       '중구 대둔', '중구 태평', '중구 평촌', '중구 대종', ' ', '중구 방아', '중구 대전', '중구 동서',
       '중구 대흥', '수침로 9', '중앙로 1', '대전 대덕', '비공개시설', '\n대전광역', '대전시 대',
       '대전시 동', '외국 58', '외국 61', '외국 48', ' 대전광역', '대전시&n', '충청북도 ',
       '대전 동구', '대전 유성'], dtype=object)

* 대전광역시, 대전시, 대전 등 존재
* 제거하지 못한 특수기호들도 보임

#### `&nbsp;` 제거

In [25]:
# 대전시&n 가 뭐지?
df_space[df_space['주소'].str.contains('대전시&n')] 

구  구분      시설명                              주소  \
14859  서구  학교  대전내동중학교  대전시&nbsp;서구&nbsp;배재로&nbsp;263    

                                     번지주소  
14859  대전시&nbsp;서구&nbsp;배재로&nbsp;263 (변동)

In [26]:
# &nbsp; 17개 존재
df_space[df_space['주소'].str.contains('&nbsp;')].shape

(17, 5)

In [27]:
df_space["주소"] = df_space["주소"].str.replace(pat='&nbsp;', repl=r'', regex=True)

In [28]:
df_space[df_space['주소'].str.contains('&nbsp;')].shape # 제거완료

(0, 5)

#### 대전광역시 & 대전시 & 대전 제거

In [29]:
df_space["주소"] = df_space["주소"].str.replace(pat='대전광역시', repl=r'', regex=True)
df_space["주소"] = df_space["주소"].str.replace(pat='대전시', repl=r'', regex=True)
# df3["주소"] = df3["주소"].str.replace(pat='대전', repl=r'', regex=True) 
# 대전을 삭제하면 동대전로 대전로 등 같이 삭제됨

In [30]:
df_space['주소'].str[:4].unique()

array([' 중구 ', '대전 중', '대저너시', ' 충무로', '중구 서', ' 동구 ', '대전 서', ' 중구목',
       '서울시 ', ' 중구', '보문산로', '선화로4', '대전천서', '부용로4', '당디로 ', '단재로 ',
       '유천로1', '충무로1', '대종로 ', '사득로 ', '천근로 ', '당골길 ', '테미로 ', '범골로 ',
       '성산로 ', '성산로1', '오류로 ', '충무로 ', '대종로5', '계백로1', '봉소루로', '돌다리로',
       '산서로 ', '대흥로 ', '문창로 ', '태평로1', '계백로 ', '대둔산로', '문화로1', '중앙로 ',
       '중구 오', '중구 계', '중구 중', '중구 대', '중구 태', '중구 평', ' ', '중구 방',
       '중구 동', '수침로 ', ' 대덕구', '대전 대', '비공개시', '\n 대덕', ' 비래동', ' 송촌북',
       ' 벚꽃길', ' 계족로', ' 신탄진', ' 동춘당', ' 비래서', ' 오정로', ' 엣신탄', ' 덕암로',
       ' 석봉북', ' 계족산', ' 대전로', ' 한밭대', ' 동산초', ' 우암동', ' 대덕대', ' 중리북',
       ' 서구 ', '외국 5', '외국 6', '외국 4', '  서구', '서구가장', '서구도산', '서구배재',
       '서구원도', '서구벌곡', '서구둔산', ' 서구유', '서구신갈', '서구기성', '서구탄방', '서구동서',
       '서구정림', ' 서구월', '서구대덕', '서구관저', '서구복수', '충청북도', ' 서구조', ' 서구흑',
       '대전 동', '?동구?', ' 동구?', ' 동구옥', ' 동구백', ' 유성구', ' 전민로', '대전 유'],
      dtype=object)

* 대저너시, '  ', \n, ?, 서울시, 외국 등 확인
* 대전천서, 대전 서, 대전 중, 대전 대, 대전로, 대전 유, 대전 동 등 `대전` + `도로명` 형태 확인
    * 대전 뒤에 일괄적으로 공백 삽입 후 다시 형태 확인

#### 대저너시, '  ', \n, ?, 서울시 등 확인

In [31]:
# \n 제거
df_space[df_space['주소'].str.contains('\n')]

구      구분              시설명                                주소  \
11914  대덕구      청사            수질연구소                    \n 대덕구 송촌로 52    
29172   동구  사회복지시설             참좋은집            동구 성동로74, 102동 201호\n    
29173   동구  사회복지시설             소망의집          동구 계족로 409번길 33, 302호\n    
29174   동구  사회복지시설           신흥마을의집           동구 옥천로 38, 101동 704호\n    
29175   동구  사회복지시설            샤인하우스          동구 신기로 100, 607동104호 \n    
29187   동구  사회복지시설         한빛지역아동센터       동구 옥천로 180번길15,\n리라빌딩 301호    
29198   동구  사회복지시설      파랑새휴먼지역아동센터          동구 한밭대로1237번길\n45-14 ,2층   
29267   동구  사회복지시설           주사랑 쉼터      동구 대전로 542번길 121, 316/301\n    
29268   동구  사회복지시설           대전밀알의집                동구 용운로 156번길 15\n    
29271   동구  사회복지시설              해달별        동구 계족로 489번길 73,\n101-101    
29273   동구  사회복지시설               정원       동구 대전로 340번길 20, \n105-105    
29275   동구  사회복지시설            주사랑의집   동구 대전로 542번길 121, 313동 103호 \n    
29276   동구  사회복지시설              우리들                  동구 학고개로 74-27\n    
29286   동구  사회복지시설  대전광역시수어통역센터지역본부           동구 중앙로200번길66\n2층 205호    

                                                 번지주소  
11914                            \n대전광역시 대덕구 송촌로 52    
29172         대전광역시 동구 성동로74, 102동 201호\n(자양동, 동아아파트)  
29173        대전광역시 동구 계족로 409번길 33, 302호\n(홍도동, 공원빌라)  
29174       대전광역시 동구 옥천로 38, 101동 704호\n(신흥동,신흥마을아파트)  
29175    대전광역시 동구 신기로 100, 607동104호 \n(가오동, 은어송마을아파트)  
29187          대전광역시 동구 옥천로 180번길15,\n리라빌딩 301호(판암1동)  
29198              대전광역시 동구 한밭대로1237번길\n45-14(용전동),2층  
29267     대전광역시 동구 대전로 542번길 121, 316/301\n(천동, 위드윌@)  
29268                    대전광역시 동구 용운로 156번길 15\n(용운동)  
29271       대전광역시 동구 계족로 489번길 73,\n101-101(용전동, 한숲@)  
29273      대전광역시 동구 대전로 340번길 20, \n105-105(대성동, 삼익@)  
29275  대전광역시 동구 대전로 542번길 121, 313동 103호 \n(천동, 위드휠@)  
29276                      대전광역시 동구 학고개로 74-27\n(이사동)  
29286                대전광역시 동구 중앙로200번길66\n2층 205호(원동)

In [32]:
# \n 제거
df_space["주소"] = df_space["주소"].str.replace(pat='\n', repl=r' ', regex=True)

In [33]:
# ? 가 들어있는 행 확인
df_space[df_space['주소'].str.contains('\?')]

구            구분               시설명  \
624     중구  어린이운송용 승합자동차        용인대온누리태권도장   
1122    중구        실내체육시설            제일당구클럽   
16334   서구            학원            경희음악학원   
16420   서구            학원   튼튼영어마스터클럽대전관저학원   
16487   서구            학원      관저디자인애니톡미술학원   
16681   서구            학원          둔산이화무용학원   
16801   서구            학원              대치학원   
16877   서구            학원     다빈치영재입시센터둔산학원   
16878   서구            학원            메이몽어학원   
16966   서구            학원            엠투수학학원   
17000   서구            학원         조형가우디미술학원   
17002   서구            학원         둔산케이툰만화학원   
17037   서구            학원            초록음악학원   
17140   서구            학원       보이스타운실용음악학원   
28617   동구  어린이운송용 승합자동차        꿈나무유치원(사립)   
28619   동구  어린이운송용 승합자동차  대전보건대학교부속유치원(사립)   
28620   동구  어린이운송용 승합자동차        동대전유치원(사립)   
28621   동구  어린이운송용 승합자동차         동심유치원(사립)   
28625   동구  어린이운송용 승합자동차         성모유치원(사립)   
28626   동구  어린이운송용 승합자동차     (휴원)송산유치원(사립)   
28627   동구  어린이운송용 승합자동차      아이들세상유치원(사립)   
28628   동구  어린이운송용 승합자동차         예지유치원(사립)   
28629   동구  어린이운송용 승합자동차         용운유치원(사립)   
28630   동구  어린이운송용 승합자동차         천사유치원(사립)   
28631   동구  어린이운송용 승합자동차       푸른나라유치원(사립)   
28632   동구  어린이운송용 승합자동차         하람유치원(사립)   
41169  유성구            학원             페르마학원   
41300  유성구            학원          메이플베어어학원   
42798  유성구      어린이운송용차량         용인대정심태권도2   

                                                      주소  \
624               중구 태평동 422-7?; 버드내 상가 301호, 용인대온누리태권도장   
1122                                   중구 목중로 50, 402호 ?   
16334                             서구 갈마로 276? 2층 세연음악학원    
16420                                  서구 관저동로 66? 401호    
16487                               서구 관저중로95번길 104? 4층    
16681                                 서구 둔산남로9번길 43? 4층    
16801                              서구 둔산로137번길 17? 4층일부    
16877                              서구 둔산중로58번길 25? 2층전체    
16878                              서구 둔산중로58번길 25? 5층일부    
16966                                   서구 문예로 65? 4층일부    
17000                                     서구 문정로 84? 3층    
17002                                     서구 문정로 90? 3층    
17037                                    서구 복수남로 56? 3층    
17140                                    서구 월평북로 81? 6층    
28617                                       ?동구?옻밭2길 77    
28619                                        ?동구?충정로?21    
28620                                       ?동구?대전로 646    
28621                                동구?새울로?109번길?26-24    
28625                                ?동구?동서대로?1692번길?64    
28626                                   ?동구?동산초교로?15-15    
28627                                   ?동구?산내로?1396-34    
28628                                  ?동구?비래서로 42번길?58    
28629                                       ?동구?백룡로?104    
28630                                      ?동구?동구청로?206    
28631                                   ?동구?동부로?73번길?80    
28632                                        ?동구?삼정로?46    
41169                               유성구 테크노4로 115? 501호    
41300   유성구 엑스포로123번길 65-38? 스마트시티주상복합아파트 상가205동 301~...   
42798                                     유성구 유성대로 1728?   

                                                    번지주소  
624         대전광역시 중구 태평동 422-7?; 버드내 상가 301호, 용인대온누리태권도장  
1122                  대전광역시 중구 목중로 50, 402호(중촌동,제일프라자) ?  
16334                    대전광역시 서구 갈마로 276? 2층 세연음악학원(내동)  
16420                        대전광역시 서구 관저동로 66? 401호(관저동)  
16487                     대전광역시 서구 관저중로95번길 104? 4층(관저동)  
16681                       대전광역시 서구 둔산남로9번길 43? 4층(둔산동)  
16801                    대전광역시 서구 둔산로137번길 17? 4층일부(둔산동)  
16877                    대전광역시 서구 둔산중로58번길 25? 2층전체(둔산동)  
16878                    대전광역시 서구 둔산중로58번길 25? 5층일부(둔산동)  
16966                         대전광역시 서구 문예로 65? 4층일부(둔산동)  
17000                           대전광역시 서구 문정로 84? 3층(탄방동)  
17002                           대전광역시 서구 문정로 90? 3층(탄방동)  
17037                          대전광역시 서구 복수남로 56? 3층(복수동)  
1

In [34]:
df_space["주소"] = df_space["주소"].str.replace(pat='\?', repl=r' ', regex=True) # ? 제거

In [35]:
df_space[df_space['주소'].str.contains('대저너시')]

구            구분    시설명                                       주소  \
422  중구  어린이운송용 승합자동차  오현유치원  대저너시 중구 당디로6번길 69, 오현유치원, 오현어린이집, 오현태권도   

                                        번지주소  
422  대저너시 중구 당디로6번길 69, 오현유치원, 오현어린이집, 오현태권도

In [36]:
# 대저너시 제거
df_space["주소"] = df_space["주소"].str.replace(pat='대저너시', repl=r' ', regex=True)

In [37]:
df_space[df_space['주소'].str.contains('서울시')]

구            구분      시설명                     주소                   번지주소
785  중구  어린이운송용 승합자동차  오즈킨더유치원  서울시 마포구 신촌로 170, 820호  서울시 마포구 신촌로 170, 820호

In [38]:
# 서울시 행 삭제
df_space = df_space.drop([785], axis=0)

In [39]:
# 외국이 뭐지?
df_space[df_space['주소'].str.contains('외국')]

구                 구분             시설명                              주소  \
571     중구       어린이운송용 승합자동차         삼육외국어학원   대전 중구 대사동 248-349번지 , 삼육외국어학원   
773     중구       어린이운송용 승합자동차         삼육외국어학원     중구 계룡로921번길 8, 1~3층 삼육외국어학원   
777     중구       어린이운송용 승합자동차         삼육외국어학원     중구 계룡로921번길 8, 1~3층 삼육외국어학원   
811     중구       어린이운송용 승합자동차         삼육외국어학원      중구 계룡로921번길 8,1~3층 삼육외국어학원   
819     중구       어린이운송용 승합자동차         삼육외국어학원   중구 계룡로921번길 8, 1~3층 , 삼육외국어학원   
13109   서구  연면적 1천제곱미터 이상 건축물             NaN                        외국 588-1   
13378   서구  연면적 1천제곱미터 이상 건축물             NaN                       외국 610-10   
13381   서구  연면적 1천제곱미터 이상 건축물             NaN                        외국 484-0   
13383   서구  연면적 1천제곱미터 이상 건축물             NaN                        외국 610-0   
13384   서구  연면적 1천제곱미터 이상 건축물             NaN                        외국 610-7   
13385   서구  연면적 1천제곱미터 이상 건축물              1동                       외국 610-11   
37257  유성구                음식점  제이제이케터링대전외국인학교    유성구 용산동 709번지 대전외국인학교, 지상1층    

                                         번지주소  
571        대전 중구 대사동 248-349번지(1~3층), 삼육외국어학원  
773    대전광역시 중구 계룡로921번길 8, 1~3층(대사동) 삼육외국어학원  
777    대전광역시 중구 계룡로921번길 8, 1~3층(대사동) 삼육외국어학원  
811       대전시 중구 계룡로921번길 8,1~3층(대사동) 삼육외국어학원  
819     대전시 중구 계룡로921번길 8, 1~3층(대사동), 삼육외국어학원  
13109                              외국   588-1  
13378                             외국   610-10  
13381                              외국   484-0  
13383                              외국   610-0  
13384                              외국   610-7  
13385                             외국   610-11  
37257      대전광역시 유성구 용산동 709번지 대전외국인학교, 지상1층

In [40]:
# 외국으로 시작하는 지번 제거
df_space = df_space.drop([13109, 13378, 13381, 13383, 13384, 13385], axis=0) # 외국 삭제

In [41]:
df_space.reset_index(drop=True, inplace=True)
df_space.loc[13109]

구                        서구
구분        연면적 1천제곱미터 이상 건축물
시설명                     NaN
주소            서구 둔산동 1335-0
번지주소    대전광역시 서구 둔산동 1335-0
Name: 13109, dtype: object

In [42]:
# 다시 주소 형태 확인
df_space['주소'].str[:4].unique()

array([' 중구 ', '대전 중', '  중구', ' 충무로', '중구 서', ' 동구 ', '대전 서', ' 중구목',
       ' 중구', '보문산로', '선화로4', '대전천서', '부용로4', '당디로 ', '단재로 ', '유천로1',
       '충무로1', '대종로 ', '사득로 ', '천근로 ', '당골길 ', '테미로 ', '범골로 ', '성산로 ',
       '성산로1', '오류로 ', '충무로 ', '대종로5', '계백로1', '봉소루로', '돌다리로', '산서로 ',
       '대흥로 ', '문창로 ', '태평로1', '계백로 ', '대둔산로', '문화로1', '중앙로 ', '중구 오',
       '중구 계', '중구 중', '중구 대', '중구 태', '중구 평', ' ', '중구 방', '중구 동',
       '수침로 ', ' 대덕구', '대전 대', '비공개시', '  대덕', ' 비래동', ' 송촌북', ' 벚꽃길',
       ' 계족로', ' 신탄진', ' 동춘당', ' 비래서', ' 오정로', ' 엣신탄', ' 덕암로', ' 석봉북',
       ' 계족산', ' 대전로', ' 한밭대', ' 동산초', ' 우암동', ' 대덕대', ' 중리북', ' 서구 ',
       '  서구', '서구가장', '서구도산', '서구배재', '서구원도', '서구벌곡', '서구둔산', ' 서구유',
       '서구신갈', '서구기성', '서구탄방', '서구동서', '서구정림', ' 서구월', '서구대덕', '서구관저',
       '서구복수', '충청북도', ' 서구조', ' 서구흑', '대전 동', ' 동구옥', ' 동구백', ' 유성구',
       ' 전민로', '대전 유'], dtype=object)

* 앞자리가 공백으로 시작하는 것 확인 -> 앞 공백 제거
* 대전천서, 대전 서, 대전 중, 대전 대, 대전로, 대전 유, 대전 동 등 `대전` + `도로명` 형태 확인
    * 대전 뒤에 일괄적으로 공백 삽입 후 다시 형태 확인

In [43]:
# 대전 뒤에 공백 삽입
df_space['주소'] = df_space['주소'].str.replace('대전', '대전 ')
df_space[df_space['주소'].str.contains('대전')]

구            구분       시설명                                       주소  \
337     중구  어린이운송용 승합자동차    강남어린이집                  대전  중구 계룡로873-11 강남어린이집   
338     중구  어린이운송용 승합자동차  해바라기어린이집             대전  중구 목동 목양마을@ 단지내 해바라기어린이집   
339     중구  어린이운송용 승합자동차     오현유치원                   대전  중구 문화동 780번지 오현유치원   
342     중구  어린이운송용 승합자동차    꿈터어린이집              대전  중구 부사동 100번지 10호 꿈터어린이집   
343     중구  어린이운송용 승합자동차   랄랄라어린이집  대전  중구 태평동 393번지 삼부아파트 410동 12호 랄랄라어린이집   
...    ...           ...       ...                                      ...   
44247  유성구       어린이놀이시설   호산어린이공원                          대전  유성구 학하동 730   
44248  유성구       어린이놀이시설   별빛어린이공원                        대전  유성구 학하동 715-1   
44249  유성구       어린이놀이시설  학무정어린이공원                        대전  유성구 학하동 763-4   
44250  유성구       어린이놀이시설  영당말어린이공원                        대전  유성구 학하동 778-1   
44251  유성구       어린이놀이시설   월봉어린이공원                        대전  유성구 학하동 783-3   

                                         번지주소  
337                    대전 중구 계룡로873-11 강남어린이집  
338               대전 중구 목동 목양마을@ 단지내 해바라기어린이집  
339                     대전 중구 문화동 780번지 오현유치원  
342                대전 중구 부사동 100번지 10호 꿈터어린이집  
343    대전 중구 태평동 393번지 삼부아파트 410동 12호 랄랄라어린이집  
...                                       ...  
44247                          대전 유성구 학하동 730  
44248                        대전 유성구 학하동 715-1  
44249                        대전 유성구 학하동 763-4  
44250                        대전 유성구 학하동 778-1  
44251                        대전 유성구 학하동 783-3  

[2153 rows x 5 columns]

In [45]:
df_space['주소'].str[:4].unique()

array([' 중구 ', '대전  ', '  중구', ' 충무로', '중구 서', ' 동구 ', ' 중구목', ' 중구',
       '보문산로', '선화로4', '대전 천', '부용로4', '당디로 ', '단재로 ', '유천로1', '충무로1',
       '대종로 ', '사득로 ', '천근로 ', '당골길 ', '테미로 ', '범골로 ', '성산로 ', '성산로1',
       '오류로 ', '충무로 ', '대종로5', '계백로1', '봉소루로', '돌다리로', '산서로 ', '대흥로 ',
       '문창로 ', '태평로1', '계백로 ', '대둔산로', '문화로1', '중앙로 ', '중구 오', '중구 계',
       '중구 중', '중구 대', '중구 태', '중구 평', ' ', '중구 방', '중구 동', '수침로 ',
       ' 대덕구', '비공개시', '  대덕', ' 비래동', ' 송촌북', ' 벚꽃길', ' 계족로', ' 신탄진',
       ' 동춘당', ' 비래서', ' 오정로', ' 엣신탄', ' 덕암로', ' 석봉북', ' 계족산', ' 대전 ',
       ' 한밭대', ' 동산초', ' 우암동', ' 대덕대', ' 중리북', ' 서구 ', '  서구', '서구가장',
       '서구도산', '서구배재', '서구원도', '서구벌곡', '서구둔산', ' 서구유', '서구신갈', '서구기성',
       '서구탄방', '서구동서', '서구정림', ' 서구월', '서구대덕', '서구관저', '서구복수', '충청북도',
       ' 서구조', ' 서구흑', ' 동구옥', ' 동구백', ' 유성구', ' 전민로'], dtype=object)

#### 다시 중복된 공백 제거

In [46]:
# 중복 공백 확인
print(df_space['주소'].str.contains('  ').sum()) # 스페이스바 두번
print(df_space['주소'].str.contains('   ').sum()) # 스페이스바 세번
print(df_space['주소'].str.contains('    ').sum()) # 스페이스바 네번

701
2
0


In [48]:
# 전부 0이 될때까지 여러번 실행
df_space["주소"] = df_space["주소"].str.replace(pat='  ', repl=r' ', regex=True)
print(df_space['주소'].str.contains('  ').sum()) # 스페이스바 두번
print(df_space['주소'].str.contains('   ').sum()) # 스페이스바 세번

0
0


In [49]:
df_space['주소'].str[:4].unique()

array([' 중구 ', '대전 중', ' 충무로', '중구 서', ' 동구 ', '대전 서', ' 중구목', ' 중구',
       '보문산로', '선화로4', '대전 천', '부용로4', '당디로 ', '단재로 ', '유천로1', '충무로1',
       '대종로 ', '사득로 ', '천근로 ', '당골길 ', '테미로 ', '범골로 ', '성산로 ', '성산로1',
       '오류로 ', '충무로 ', '대종로5', '계백로1', '봉소루로', '돌다리로', '산서로 ', '대흥로 ',
       '문창로 ', '태평로1', '계백로 ', '대둔산로', '문화로1', '중앙로 ', '중구 오', '중구 계',
       '중구 중', '중구 대', '중구 태', '중구 평', ' ', '중구 방', '중구 동', '수침로 ',
       ' 대덕구', '대전 대', '비공개시', ' 비래동', ' 송촌북', ' 벚꽃길', ' 계족로', ' 신탄진',
       ' 동춘당', ' 비래서', ' 오정로', ' 엣신탄', ' 덕암로', ' 석봉북', ' 계족산', ' 대전 ',
       ' 한밭대', ' 동산초', ' 우암동', ' 대덕대', ' 중리북', ' 서구 ', '서구가장', '서구도산',
       '서구배재', '서구원도', '서구벌곡', '서구둔산', ' 서구유', '서구신갈', '서구기성', '서구탄방',
       '서구동서', '서구정림', ' 서구월', '서구대덕', '서구관저', '서구복수', '충청북도', ' 서구조',
       ' 서구흑', '대전 동', ' 동구옥', ' 동구백', ' 유성구', ' 전민로', '대전 유'],
      dtype=object)

#### 맨 앞 공백제거

In [51]:
df_space["주소"] = df_space["주소"].str.lstrip()

In [52]:
df_space['주소'].str[:3].unique()

array(['중구 ', '대전 ', '충무로', '동구 ', '중구목', '중구', '보문산', '선화로', '부용로',
       '당디로', '단재로', '유천로', '대종로', '사득로', '천근로', '당골길', '테미로', '범골로',
       '성산로', '오류로', '계백로', '봉소루', '돌다리', '산서로', '대흥로', '문창로', '태평로',
       '대둔산', '문화로', '중앙로', '', '수침로', '대덕구', '비공개', '비래동', '송촌북', '벚꽃길',
       '계족로', '신탄진', '동춘당', '비래서', '오정로', '엣신탄', '덕암로', '석봉북', '계족산',
       '한밭대', '동산초', '우암동', '대덕대', '중리북', '서구 ', '서구가', '서구도', '서구배',
       '서구원', '서구벌', '서구둔', '서구유', '서구신', '서구기', '서구탄', '서구동', '서구정',
       '서구월', '서구대', '서구관', '서구복', '충청북', '서구조', '서구흑', '동구옥', '동구백',
       '유성구', '전민로'], dtype=object)

In [53]:
df_space

구    구분      시설명               주소                        번지주소
0       중구    학교   굿모닝유치원  중구 어덕마을로 84-12   대전광역시 중구 어덕마을로 84-12 (중촌동)
1       중구    학교    금동유치원  중구 태평로152번길 63   대전광역시 중구 태평로152번길 63 (태평동)
2       중구    학교   꿈내리유치원       중구 당디로 45        대전광역시 중구 당디로 45 (산성동)
3       중구    학교  남대전고등학교  중구 보문로111번길 63   대전광역시 중구 보문로111번길 63 (대사동)
4       중구    학교   대전고등학교      중구 대흥로 110       대전광역시 중구 대흥로 110 (대흥동)
...    ...   ...      ...              ...                         ...
45594  유성구  복합용도       식당     유성구 학하동676-1          대전광역시 유성구 학하동676-1
45595  유성구  복합용도      NaN    유성구 학하동713-12         대전광역시 유성구 학하동713-12
45596  유성구  복합용도      NaN     유성구 학하동716-7          대전광역시 유성구 학하동716-7
45597  유성구  복합용도      NaN     유성구 화암동157-6          대전광역시 유성구 화암동157-6
45598  유성구  복합용도      NaN     유성구 화암동63-10          대전광역시 유성구 화암동63-10

[45599 rows x 5 columns]

## 구 분리

대전시는 유성구, 대덕구, 중구, 서구, 동구 다섯가지 지역구로 구성됨  
주소에서 구를 분리하는 작업
 * 구 뒤에 공백 삽업


In [54]:
df_gu = df_space.copy()

In [55]:
df_gu['구'].unique()

array(['중구', '대덕구', '서구', '동구', '유성구'], dtype=object)

#### 구 뒤에 공백 삽입

In [56]:
# 구 뒤에 공백 삽입
df_gu['주소'] = df_gu['주소'].str.replace('중구', '중구 ')
df_gu['주소'] = df_gu['주소'].str.replace('대덕구', '대덕구 ')
df_gu['주소'] = df_gu['주소'].str.replace('서구', '서구 ')
df_gu['주소'] = df_gu['주소'].str.replace('동구', '동구 ')
df_gu['주소'] = df_gu['주소'].str.replace('유성구', '유성구 ')

#### `**구` 제거

In [57]:
df_gu["주소"] = df_gu["주소"].str.replace(pat='중구 ', repl=r'', regex=True)
df_gu["주소"] = df_gu["주소"].str.replace(pat='대덕구 ', repl=r'', regex=True)
df_gu["주소"] = df_gu["주소"].str.replace(pat='서구 ', repl=r'', regex=True)
df_gu["주소"] = df_gu["주소"].str.replace(pat='동구 ', repl=r'', regex=True)
df_gu["주소"] = df_gu["주소"].str.replace(pat='유성구 ', repl=r'', regex=True)

In [58]:
df_gu[df_gu['주소'].str.contains('중구|대덕구|서구|동구|유성구')]

Empty DataFrame
Columns: [구, 구분, 시설명, 주소, 번지주소]
Index: []

In [59]:
# 공백제거
df_gu["주소"] = df_gu["주소"].str.replace(pat='  ', repl=r' ', regex=True)
df_gu["주소"] = df_gu["주소"].str.replace(pat='  ', repl=r' ', regex=True)

In [60]:
print(df_gu['주소'].str.contains('  ').sum()) # 스페이스바 두번
print(df_gu['주소'].str.contains('   ').sum()) # 스페이스바 세번
print(df_gu['주소'].str.contains('    ').sum()) # 스페이스바 네번

0
0
0


In [61]:
# 맨 앞에 공백 제거
df_gu["주소"] = df_gu["주소"].str.lstrip()

In [62]:
df_gu['주소'].str[:2].unique()

array(['어덕', '태평', '당디', '보문', '대흥', '서문', '계룡', '송리', '목동', '솔밭', '수도',
       '문화', '목중', '유등', '문창', '대종', '대둔', '선화', '부사', '오류', '모암', '계백',
       '수침', '유천', '사득', '산성', '남달', '학고', '동서', '평촌', '천근', '중앙', '충무',
       '중교', '과례', '대전', '사정', '중촌', '석교', '돌다', '용두', '산서', '문로', '테미',
       '동부', '호동', '', '봉소', '안영', '뿌리', '인창', '목척', '부용', '단재', '당골',
       '범골', '성산', '신기', '대사', '침산', '방아', '우암', '운남', '정생', '잔디', '가재',
       '한남', '신탄', '홍도', '계족', '중리', '덕암', '동춘', '대덕', '비래', '한밭', '석봉',
       '옛신', '송촌', '오정', '신일', '대청', '아리', '선비', '대화', '문평', '동심', '벚꽃',
       '대오', '생산', '신상', '쌍청', '목상', '법동', '갑천', '상서', '늘봄', '장동', '산디',
       '남경', '새말', '청자', '태양', '산업', '회덕', '방두', '동성', '비공', '미호', '와동',
       '신대', '읍내', '연축', '삼정', '용호', '동산', '청림', '무지', '엣신', '흥룡', '다사',
       '복수', '용문', '둔산', '도마', '탄방', '변동', '관저', '괴정', '월평', '도안', '갈마',
       '만년', '가수', '정림', '내동', '가장', '오동', '흑석', '장안', '우명', '매노', '산직',
       '괴곡', '원정', '원도', '청사', '구봉', '신갈', '배재', '중반'

In [63]:
df_gu

구    구분      시설명            주소                        번지주소
0       중구    학교   굿모닝유치원  어덕마을로 84-12   대전광역시 중구 어덕마을로 84-12 (중촌동)
1       중구    학교    금동유치원  태평로152번길 63   대전광역시 중구 태평로152번길 63 (태평동)
2       중구    학교   꿈내리유치원       당디로 45        대전광역시 중구 당디로 45 (산성동)
3       중구    학교  남대전고등학교  보문로111번길 63   대전광역시 중구 보문로111번길 63 (대사동)
4       중구    학교   대전고등학교      대흥로 110       대전광역시 중구 대흥로 110 (대흥동)
...    ...   ...      ...           ...                         ...
45594  유성구  복합용도       식당      학하동676-1          대전광역시 유성구 학하동676-1
45595  유성구  복합용도      NaN     학하동713-12         대전광역시 유성구 학하동713-12
45596  유성구  복합용도      NaN      학하동716-7          대전광역시 유성구 학하동716-7
45597  유성구  복합용도      NaN      화암동157-6          대전광역시 유성구 화암동157-6
45598  유성구  복합용도      NaN      화암동63-10          대전광역시 유성구 화암동63-10

[45599 rows x 5 columns]

## 도로명주소 전처리 시도

도로명주소는 

`행정구역` `도로명건물번호` `,` `상세주소` 로 구분됨.  
첫 `,`를 기준으로 나누면 도로명주소의 훼손없이 분리 가능

In [64]:
df_split = df_gu.copy()
df_split['세부주소'] = np.nan
df_split

구    구분      시설명            주소                        번지주소  세부주소
0       중구    학교   굿모닝유치원  어덕마을로 84-12   대전광역시 중구 어덕마을로 84-12 (중촌동)   NaN
1       중구    학교    금동유치원  태평로152번길 63   대전광역시 중구 태평로152번길 63 (태평동)   NaN
2       중구    학교   꿈내리유치원       당디로 45        대전광역시 중구 당디로 45 (산성동)   NaN
3       중구    학교  남대전고등학교  보문로111번길 63   대전광역시 중구 보문로111번길 63 (대사동)   NaN
4       중구    학교   대전고등학교      대흥로 110       대전광역시 중구 대흥로 110 (대흥동)   NaN
...    ...   ...      ...           ...                         ...   ...
45594  유성구  복합용도       식당      학하동676-1          대전광역시 유성구 학하동676-1   NaN
45595  유성구  복합용도      NaN     학하동713-12         대전광역시 유성구 학하동713-12   NaN
45596  유성구  복합용도      NaN      학하동716-7          대전광역시 유성구 학하동716-7   NaN
45597  유성구  복합용도      NaN      화암동157-6          대전광역시 유성구 화암동157-6   NaN
45598  유성구  복합용도      NaN      화암동63-10          대전광역시 유성구 화암동63-10   NaN

[45599 rows x 6 columns]

### 콤마가 있는 값 분리

In [65]:
df_split_com = df_split[df_split['주소'].str.contains(',')]
df_split_ncom = df_split[~df_split['주소'].str.contains(',')]
print(df_split_com.shape)
print(df_split_ncom.shape)

(12360, 6)
(33239, 6)


In [66]:
df_split_com 

구        구분              시설명                               주소  \
95      중구  학교교과교습학원            중앙독서실                    목중로26번길 , 83    
96      중구  학교교과교습학원      예녹피아노음악종합학원         당디로 45 , 201호,203호,204호    
97      중구  학교교과교습학원           단비종합학원                보문로 180 1층,2층,3층    
98      중구  학교교과교습학원             국제학원                   대둔산로 437 , 3층    
100     중구  학교교과교습학원           두루음악학원                당디로6번길 76 , 403호    
...    ...       ...              ...                              ...   
44528  유성구    지역아동센터      호산나반석지역아동센터                  지족로 218-33, 2층    
44529  유성구    지역아동센터  학교밖 청소년지원센터 꿈드림                  유성대로 798-1, 2층    
44531  유성구    만화대여업소          놀숲 충남대점                    한밭대로 400, 3층    
44542  유성구      숙박업소    호텔스카이파크 대전1호점            테크노중앙로 161, 호텔 스카이파크    
44550  유성구       목욕장        한진대중찜질목욕탕  봉명동 535번지 5호 유성한진리조트오피스텔지하1,2층    

                                                번지주소  세부주소  
95               대전광역시 중구 목중로26번길  , 83 (중촌동, 현대아파트)   NaN  
96     대전광역시 중구 당디로 45 , 201호,203호,204호 (산성동, 우성아파트)   NaN  
97                   대전광역시 중구 보문로 180 1층,2층,3층 (대흥동)   NaN  
98                      대전광역시 중구 대둔산로 437 , 3층 (산성동)   NaN  
100                  대전광역시 중구 당디로6번길 76 , 403호 (문화동)   NaN  
...                                              ...   ...  
44528                  대전광역시 유성구 지족로 218-33, 2층(지족동)   NaN  
44529                  대전광역시 유성구 유성대로 798-1, 2층(장대동)   NaN  
44531                    대전광역시 유성구 한밭대로 400, 3층 (궁동)   NaN  
44542           대전광역시 유성구 테크노중앙로 161, 호텔 스카이파크 (용산동)   NaN  
44550     대전광역시 유성구 봉명동 535번지 5호  유성한진리조트오피스텔지하1,2층    NaN  

[12360 rows x 6 columns]

In [67]:
# 첫번째 콤마를 기준으로 분리
a = df_split_com['주소'].str.split(',', expand=True, n=1)

In [68]:
a

0                 1
95                        목중로26번길                83 
96                         당디로 45    201호,203호,204호 
97                      보문로 180 1층            2층,3층 
98                       대둔산로 437                3층 
100                     당디로6번길 76              403호 
...                            ...               ...
44528                   지족로 218-33               2층 
44529                   유성대로 798-1               2층 
44531                     한밭대로 400               3층 
44542                   테크노중앙로 161         호텔 스카이파크 
44550  봉명동 535번지 5호 유성한진리조트오피스텔지하1               2층 

[12360 rows x 2 columns]

In [69]:
# df_split_com에 a 값 대체
df_split_com['주소'] = a[0].copy()
df_split_com['세부주소'] = a[1].copy()

C:\Users\Helpe\AppData\Local\Temp\ipykernel_6112\3182168813.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_split_com['주소'] = a[0].copy()
C:\Users\Helpe\AppData\Local\Temp\ipykernel_6112\3182168813.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_split_com['세부주소'] = a[1].copy()


In [70]:
df_split_com

구        구분              시설명                           주소  \
95      중구  학교교과교습학원            중앙독서실                     목중로26번길    
96      중구  학교교과교습학원      예녹피아노음악종합학원                      당디로 45    
97      중구  학교교과교습학원           단비종합학원                   보문로 180 1층   
98      중구  학교교과교습학원             국제학원                    대둔산로 437    
100     중구  학교교과교습학원           두루음악학원                   당디로6번길 76    
...    ...       ...              ...                          ...   
44528  유성구    지역아동센터      호산나반석지역아동센터                   지족로 218-33   
44529  유성구    지역아동센터  학교밖 청소년지원센터 꿈드림                   유성대로 798-1   
44531  유성구    만화대여업소          놀숲 충남대점                     한밭대로 400   
44542  유성구      숙박업소    호텔스카이파크 대전1호점                   테크노중앙로 161   
44550  유성구       목욕장        한진대중찜질목욕탕  봉명동 535번지 5호 유성한진리조트오피스텔지하1   

                                                번지주소              세부주소  
95               대전광역시 중구 목중로26번길  , 83 (중촌동, 현대아파트)               83   
96     대전광역시 중구 당디로 45 , 201호,203호,204호 (산성동, 우성아파트)   201호,203호,204호   
97                   대전광역시 중구 보문로 180 1층,2층,3층 (대흥동)            2층,3층   
98                      대전광역시 중구 대둔산로 437 , 3층 (산성동)               3층   
100                  대전광역시 중구 당디로6번길 76 , 403호 (문화동)             403호   
...                                              ...               ...  
44528                  대전광역시 유성구 지족로 218-33, 2층(지족동)               2층   
44529                  대전광역시 유성구 유성대로 798-1, 2층(장대동)               2층   
44531                    대전광역시 유성구 한밭대로 400, 3층 (궁동)               3층   
44542           대전광역시 유성구 테크노중앙로 161, 호텔 스카이파크 (용산동)         호텔 스카이파크   
44550     대전광역시 유성구 봉명동 535번지 5호  유성한진리조트오피스텔지하1,2층                2층   

[12360 rows x 6 columns]

In [71]:
# 병합
df_road = pd.concat([df_split_com, df_split_ncom], 
                    axis=0,
                    join='outer'
                    )

df_road.reset_index(drop=True)

구        구분          시설명          주소  \
0       중구  학교교과교습학원        중앙독서실    목중로26번길    
1       중구  학교교과교습학원  예녹피아노음악종합학원     당디로 45    
2       중구  학교교과교습학원       단비종합학원  보문로 180 1층   
3       중구  학교교과교습학원         국제학원   대둔산로 437    
4       중구  학교교과교습학원       두루음악학원  당디로6번길 76    
...    ...       ...          ...         ...   
45594  유성구      복합용도           식당    학하동676-1   
45595  유성구      복합용도          NaN   학하동713-12   
45596  유성구      복합용도          NaN    학하동716-7   
45597  유성구      복합용도          NaN    화암동157-6   
45598  유성구      복합용도          NaN    화암동63-10   

                                                번지주소              세부주소  
0                대전광역시 중구 목중로26번길  , 83 (중촌동, 현대아파트)               83   
1      대전광역시 중구 당디로 45 , 201호,203호,204호 (산성동, 우성아파트)   201호,203호,204호   
2                    대전광역시 중구 보문로 180 1층,2층,3층 (대흥동)            2층,3층   
3                       대전광역시 중구 대둔산로 437 , 3층 (산성동)               3층   
4                    대전광역시 중구 당디로6번길 76 , 403호 (문화동)             403호   
...                                              ...               ...  
45594                             대전광역시 유성구 학하동676-1               NaN  
45595                            대전광역시 유성구 학하동713-12               NaN  
45596                             대전광역시 유성구 학하동716-7               NaN  
45597                             대전광역시 유성구 화암동157-6               NaN  
45598                             대전광역시 유성구 화암동63-10               NaN  

[45599 rows x 6 columns]

## Kakao api

### kakao api 검색을 위한 컬럼 생성 (구 + 주소)

In [72]:
df_kakao = df_road.copy()
df_kakao['kakao 주소'] = np.nan
df_kakao

구        구분          시설명          주소  \
95      중구  학교교과교습학원        중앙독서실    목중로26번길    
96      중구  학교교과교습학원  예녹피아노음악종합학원     당디로 45    
97      중구  학교교과교습학원       단비종합학원  보문로 180 1층   
98      중구  학교교과교습학원         국제학원   대둔산로 437    
100     중구  학교교과교습학원       두루음악학원  당디로6번길 76    
...    ...       ...          ...         ...   
45594  유성구      복합용도           식당    학하동676-1   
45595  유성구      복합용도          NaN   학하동713-12   
45596  유성구      복합용도          NaN    학하동716-7   
45597  유성구      복합용도          NaN    화암동157-6   
45598  유성구      복합용도          NaN    화암동63-10   

                                                번지주소              세부주소  \
95               대전광역시 중구 목중로26번길  , 83 (중촌동, 현대아파트)               83    
96     대전광역시 중구 당디로 45 , 201호,203호,204호 (산성동, 우성아파트)   201호,203호,204호    
97                   대전광역시 중구 보문로 180 1층,2층,3층 (대흥동)            2층,3층    
98                      대전광역시 중구 대둔산로 437 , 3층 (산성동)               3층    
100                  대전광역시 중구 당디로6번길 76 , 403호 (문화동)             403호    
...                                              ...               ...   
45594                             대전광역시 유성구 학하동676-1               NaN   
45595                            대전광역시 유성구 학하동713-12               NaN   
45596                             대전광역시 유성구 학하동716-7               NaN   
45597                             대전광역시 유성구 화암동157-6               NaN   
45598                             대전광역시 유성구 화암동63-10               NaN   

       kakao 주소  
95          NaN  
96          NaN  
97          NaN  
98          NaN  
100         NaN  
...         ...  
45594       NaN  
45595       NaN  
45596       NaN  
45597       NaN  
45598       NaN  

[45599 rows x 7 columns]

In [73]:
df_kakao['kakao 주소'] = df_kakao['구'].str.cat(df_kakao['주소'], sep=' ')

In [74]:
df_kakao.reset_index(drop=True, inplace=True)
df_kakao.head(2)

구        구분          시설명        주소  \
0  중구  학교교과교습학원        중앙독서실  목중로26번길    
1  중구  학교교과교습학원  예녹피아노음악종합학원   당디로 45    

                                            번지주소              세부주소  \
0            대전광역시 중구 목중로26번길  , 83 (중촌동, 현대아파트)               83    
1  대전광역시 중구 당디로 45 , 201호,203호,204호 (산성동, 우성아파트)   201호,203호,204호    

      kakao 주소  
0  중구 목중로26번길   
1   중구 당디로 45

In [77]:
df_kakao2 = df_kakao

df_kakao2['동'] = np.nan 
df_kakao2['위도'] = np.nan
df_kakao2['경도'] = np.nan

df_kakao2.shape

(45599, 10)

## 1차 전저리 데이터프레임 저장

In [76]:
df_kakao2.to_csv(PATH + "/03_금연구역/금연구역_장병용_1차_텍스트전처리.csv", index=False, encoding='cp949')

### kakao api 세팅

In [158]:
# rest키 발급 https://developers.kakao.com/console/app

# 병용 474aa3bdba9bd34f1560a7812fafeb9b
# 재인 08d64ac0652c8f9e1c4ea45118910c53
# 수영 616734aa8214dbf958cfaebd6bd41aec
# 태희 04e7e5d98a2c026956de6095963f28f8

rest_api_key = "474aa3bdba9bd34f1560a7812fafeb9b"

In [159]:
class KakaoLocalAPI:
    """
    Kakao Local API Controller
    """
    def __init__(self, rest_api_key):
        """
        Rest API key 초기화 및 기능 별 url 설정
        """
        
        # REST API key 설정
        self.rest_api_key = rest_api_key
        self.headers = {"Authorization": "KakaoAK {}".format(rest_api_key)}
        
        # 01 주소검색
        self.URL_01 = "https://dapi.kakao.com/v2/local/search/address.json"
        
    def search_address(self, query, analyze_type = None, page = None, size = None):
        """
        01 주소 검색
        """
        params = {"query": f"{query}"}
    
        if analyze_type != None:
            params["analyze_type"]=f"{analyze_type}"
    
        elif page != None:
            params['page'] = f"{page}"
    
        elif size != None:
            params['size'] = f"{size}"
    
        res = requests.get(self.URL_01, headers = self.headers, params = params)
        document = json.loads(res.text)
    
        return document

In [160]:
kakao = KakaoLocalAPI(rest_api_key)

In [167]:
add_list = df_kakao['kakao 주소']
add_list[0]

'중구 목중로26번길 '

In [168]:
# api 호출 확인
kakao.search_address(add_list[0]) 

{'documents': [{'address': None,
   'address_name': '대전 중구 목중로26번길',
   'address_type': 'ROAD',
   'road_address': {'address_name': '대전 중구 목중로26번길',
    'building_name': '',
    'main_building_no': '',
    'region_1depth_name': '대전',
    'region_2depth_name': '중구',
    'region_3depth_name': '중촌동',
    'road_name': '목중로26번길',
    'sub_building_no': '',
    'underground_yn': 'N',
    'x': '127.412961781981',
    'y': '36.3375130575055',
    'zone_no': ''},
   'x': '127.412961781981',
   'y': '36.3375130575055'}],
 'meta': {'is_end': True, 'pageable_count': 1, 'total_count': 1}}

### 동, 위경도 추출

In [172]:
from tqdm import tqdm
from tqdm import trange

In [176]:
# 1시간 42분 걸림
for i in tqdm(range(len(add_list))):
    try :
        df_kakao2['동'].iloc[i] = kakao.search_address(add_list[i])['documents'][0]['address']['region_3depth_name']
        df_kakao2['경도'].iloc[i] = kakao.search_address(add_list[i])['documents'][0]['x']
        df_kakao2['위도'].iloc[i] = kakao.search_address(add_list[i])['documents'][0]['y']
    except:
        pass

print(df_kakao2)

  0%|          | 0/43867 [00:00<?, ?it/s]C:\Users\JBY\AppData\Roaming\Python\Python37\site-packages\pandas\core\indexing.py:1732: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  self._setitem_single_block(indexer, value, name)
100%|██████████| 43867/43867 [1:42:27<00:00,  7.14it/s]  

         구        구분                 시설명          주소  \
0       중구  학교교과교습학원               중앙독서실    목중로26번길    
1       중구  학교교과교습학원         예녹피아노음악종합학원     당디로 45    
2       중구  학교교과교습학원              단비종합학원  보문로 180 1층   
3       중구  학교교과교습학원                국제학원   대둔산로 437    
4       중구  학교교과교습학원              두루음악학원  당디로6번길 76    
...    ...       ...                 ...         ...   
43862  유성구      복합용도             나라아이넷사옥    탑립동860-0   
43863  유성구      복합용도  파크팰리스(Park Palace)    탑립동862-0   
43864  유성구      복합용도                  농협    하기동531-1   
43865  유성구      복합용도               우성프라자    하기동531-2   
43866  유성구      복합용도                  식당    학하동676-1   

                                                번지주소              세부주소  \
0                대전광역시 중구 목중로26번길  , 83 (중촌동, 현대아파트)               83    
1      대전광역시 중구 당디로 45 , 201호,203호,204호 (산성동, 우성아파트)   201호,203호,204호    
2                    대전광역시 중구 보문로 180 1층,2층,3층 (대흥동)            2층,3층    
3                       대전광역시 중

#### kakao api 

43866건 중 11115건 실패

In [187]:
df_kakao2.isnull().sum()

구               0
구분              0
시설명             0
주소              0
번지주소            0
세부주소        31507
kakao 주소        0
동           11114
위도          11115
경도          11115
dtype: int64

## 성공한 데이터프레임과 실패한 데이터프레임 저장

In [191]:
df_fail = df_kakao2[df_kakao2['위도'].isnull()]
df_save = df_kakao2[~df_kakao2['위도'].isnull()]

print(df_fail.shape)
print(df_save.shape)

(11115, 10)
(32752, 10)


In [192]:
df_save.to_csv(PATH + "/03_금연구역/금연구역_장병용_전처리_성공(32752).csv", index=False, encoding='cp949')
df_fail.to_csv(PATH + "/03_금연구역/금연구역_장병용_전처리_실패(11115).csv", index=False, encoding='cp949')

## 실패한 주소 체크

* 총 11115 건 중 번지주소로 카카오 api 다시 호출
* 번지주소로도 검색이 안되는 경우 다시 전처리

In [101]:
# PATH = 'D:/project/데이터 분석 프로젝트/이노포스트/data'
PATH = 'G:/다른 컴퓨터/My_desktop/project/데이터 분석 프로젝트/이노포스트/data'

df_fail = pd.read_csv(PATH + '/03_금연구역/금연구역_장병용_전처리_실패(11115).csv', encoding ='cp949')
df_save = pd.read_csv(PATH + '/03_금연구역/금연구역_장병용_전처리_성공(32752).csv', encoding ='cp949')

In [85]:
df_fail2 = df_fail.copy()
df_fail2

구            구분                 시설명                     주소  \
0       중구      학교교과교습학원               중앙독서실               목중로26번길    
1       중구      학교교과교습학원             두드림음악학원          계백로1571번길 46    
2       중구  어린이운송용 승합자동차              문화어린이집  대전 보문로 207번길14 문화어린이집   
3       중구  어린이운송용 승합자동차              은혜어린이집    대전 태평동 삼부@ 410동 12호   
4       중구  어린이운송용 승합자동차             꿈나무어린이집           대전 선화로57번길 8   
...    ...           ...                 ...                    ...   
11110  유성구          복합용도             나라아이넷사옥               탑립동860-0   
11111  유성구          복합용도  파크팰리스(Park Palace)               탑립동862-0   
11112  유성구          복합용도                  농협               하기동531-1   
11113  유성구          복합용도               우성프라자               하기동531-2   
11114  유성구          복합용도                  식당               학하동676-1   

                                      번지주소         세부주소  \
0      대전광역시 중구 목중로26번길  , 83 (중촌동, 현대아파트)          83    
1         대전광역시 중구 계백로1571번길 46 , 2층 (유천동)          2층    
2       대전 중구 보문로 207번길14   문화어린이집,  문화유치원        문화유치원   
3           대전 중구 태평동 삼부@ 410동 12호, 은혜어린이집       은혜어린이집   
4         대전 중구 선화로57번길 8, 4층(선화동) 꿈나무어린이집   4층 꿈나무어린이집   
...                                    ...          ...   
11110                   대전광역시 유성구 탑립동860-0          NaN   
11111                   대전광역시 유성구 탑립동862-0          NaN   
11112                   대전광역시 유성구 하기동531-1          NaN   
11113                   대전광역시 유성구 하기동531-2          NaN   
11114                   대전광역시 유성구 학하동676-1          NaN   

                       kakao 주소    동  위도  경도  
0                   중구 목중로26번길   NaN NaN NaN  
1              중구 계백로1571번길 46   NaN NaN NaN  
2      중구 대전 보문로 207번길14 문화어린이집  NaN NaN NaN  
3        중구 대전 태평동 삼부@ 410동 12호  NaN NaN NaN  
4               중구 대전 선화로57번길 8  NaN NaN NaN  
...                         ...  ...  ..  ..  
11110              유성구 탑립동860-0  NaN NaN NaN  
11111              유성구 탑립동862-0  NaN NaN NaN  
11112              유성구 하기동531-1  NaN NaN NaN  
11113              유성구 하기동531-2  NaN NaN NaN  
11114              유성구 학하동676-1  NaN NaN NaN  

[11115 rows x 10 columns]

In [102]:
df_save.shape

(32752, 10)

In [86]:
# # rest키 발급 https://developers.kakao.com/console/app

# 병용 474aa3bdba9bd34f1560a7812fafeb9b
# 재인 08d64ac0652c8f9e1c4ea45118910c53
# 수영 616734aa8214dbf958cfaebd6bd41aec
# 태희 04e7e5d98a2c026956de6095963f28f8

rest_api_key = "04e7e5d98a2c026956de6095963f28f8"

In [87]:
class KakaoLocalAPI:
    """
    Kakao Local API Controller
    """
    def __init__(self, rest_api_key):
        """
        Rest API key 초기화 및 기능 별 url 설정
        """
        
        # REST API key 설정
        self.rest_api_key = rest_api_key
        self.headers = {"Authorization": "KakaoAK {}".format(rest_api_key)}
        
        # 01 주소검색
        self.URL_01 = "https://dapi.kakao.com/v2/local/search/address.json"
        
    def search_address(self, query, analyze_type = None, page = None, size = None):
        """
        01 주소 검색
        """
        params = {"query": f"{query}"}
    
        if analyze_type != None:
            params["analyze_type"]=f"{analyze_type}"
    
        elif page != None:
            params['page'] = f"{page}"
    
        elif size != None:
            params['size'] = f"{size}"
    
        res = requests.get(self.URL_01, headers = self.headers, params = params)
        document = json.loads(res.text)
    
        return document

In [90]:
kakao = KakaoLocalAPI(rest_api_key)

In [91]:
add_list = df_fail2['번지주소']
add_list[1]

'대전광역시 중구 계백로1571번길 46 , 2층 (유천동)'

In [92]:
kakao.search_address(add_list[1])

{'documents': [],
 'meta': {'is_end': True, 'pageable_count': 0, 'total_count': 0}}

In [97]:
# 호출에 실패한 리스트만 번지주소로 다시 돌려봄
for i in tqdm(range(len(add_list))):
    try :
        df_fail2['동'].iloc[i] = kakao.search_address(add_list[i])['documents'][0]['address']['region_3depth_name']
        df_fail2['경도'].iloc[i] = kakao.search_address(add_list[i])['documents'][0]['x']
        df_fail2['위도'].iloc[i] = kakao.search_address(add_list[i])['documents'][0]['y']
    except:
        pass

print(df_fail2.isnull().sum())

  0%|          | 0/11115 [00:00<?, ?it/s]C:\Users\Helpe\AppData\Local\Temp\ipykernel_11692\3131939199.py:4: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_fail2['동'].iloc[i] = kakao.search_address(add_list[i])['documents'][0]['address']['region_3depth_name']
C:\Users\Helpe\AppData\Local\Temp\ipykernel_11692\3131939199.py:5: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_fail2['경도'].iloc[i] = kakao.search_address(add_list[i])['documents'][0]['x']
C:\Users\Helpe\AppData\Local\Temp\ipykernel_11692\3131939199.py:6: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the cavea

구               0
구분              0
시설명             0
주소              0
번지주소            0
세부주소        10610
kakao 주소        0
동            6523
위도           6523
경도           6523
dtype: int64


카카오 api에 번지주소로 넣은 후 6523개를 여전히 인식하지 못함

### 2차 api 호출 성공/실패 데이터프레임 저장

In [100]:
df_fail3 = df_fail2[df_fail2['위도'].isnull()]
df_save3 = df_fail2[~df_fail2['위도'].isnull()]

print(df_fail3.shape)
print(df_save3.shape)

(6523, 10)
(4592, 10)


In [103]:
# 병합
df_save2= pd.concat([df_save3, df_save], 
                    axis=0,
                    join='outer'
                    )


df_save2.reset_index(drop=True)

구            구분         시설명                                 주소  \
0       중구      학교교과교습학원       중앙독서실                           목중로26번길    
1       중구  어린이운송용 승합자동차       호동유치원                           모암로 30번길   
2       중구  어린이운송용 승합자동차    그린키즈어린이집                    서문로 32 108-102    
3       중구  어린이운송용 승합자동차      꾸러기태권도                          동부로 55-58   
4       중구  어린이운송용 승합자동차      꿈내리유치원                        대전 산성동6-43    
...    ...           ...         ...                                ...   
37339  유성구           음식점         순카페    관평동 666번지 테크노벨리4단지상가 101호 102호    
37340  유성구           음식점  선비꼬마김밥 관평점  관평동 668번지 대덕테크노밸리3단지아파트상가동 101-1호   
37341  유성구           음식점        김밥도둑   관평동 669번지 대덕테크노밸리5단지 분산상가동 104호    
37342  유성구           음식점   가마치통닭 관평점         관평동 669번지 분산상가동 지상1층 101호    
37343  유성구           음식점      삼촌네불고기   관평동 672번지 대덕테크노밸리6단지아파트 상가동 106호   

                                                번지주소          세부주소  \
0                대전광역시 중구 목중로26번길  , 83 (중촌동, 현대아파트)           83    
1                대전시 중구  모암로 30번길,7-8(호동유치원, 호동어린이집)          7-8    
2      대전광역시 중구 서문로 32 108-102(문화동, 한밭우성@), 그린키즈어린이집      그린키즈어린이집   
3          대전광역시 동구 동부로 55-58, 601동 1401호(판암동 주공아파트)   601동 1401호    
4                     대전 중구 산성동6-43(우성아파트 내), 꿈내리유치원        꿈내리유치원   
...                                              ...           ...   
37339      대전광역시 유성구 관평동 666번지 테크노벨리4단지상가 101호 102호            NaN   
37340    대전광역시 유성구 관평동 668번지 대덕테크노밸리3단지아파트상가동 101-1호           NaN   
37341   대전광역시 유성구 관평동 669번지 대덕테크노밸리5단지(아)분산상가동 104호            NaN   
37342           대전광역시 유성구 관평동 669번지 분산상가동 지상1층 101호            NaN   
37343     대전광역시 유성구 관평동 672번지 대덕테크노밸리6단지아파트 상가동 106호           NaN   

                                    kakao 주소    동                위도  \
0                                중구 목중로26번길   중촌동  36.3387070586325   
1                                중구 모암로 30번길   호동  36.3025866691176   
2                         중구 서문로 32 108-102   문화동  36.3113044128371   
3                               중구 동부로 55-58  판암동   36.324283820924   
4                             중구 대전 산성동6-43    금동  36.2526076983463   
...                                      ...  ...               ...   
37339    유성구 관평동 666번지 테크노벨리4단지상가 101호 102호   관평동          36.41951   
37340  유성구 관평동 668번지 대덕테크노밸리3단지아파트상가동 101-1호  관평동         36.421222   
37341   유성구 관평동 669번지 대덕테크노밸리5단지 분산상가동 104호   관평동         36.420075   
37342         유성구 관평동 669번지 분산상가동 지상1층 101호   관평동         36.420075   
37343   유성구 관평동 672번지 대덕테크노밸리6단지아파트 상가동 106호  관평동         36.419349   

                     경도  
0      127.414436051345  
1      127.447484466761  
2      127.410149505642  
3       127.45825285256  
4      127.434591943227  
...                 ...  
37339        127.384518  
37340        127.385636  
37341        127.387263  
37342        127.387263  
37343        127.390124  

[37344 rows x 10 columns]

In [104]:
df_save2['위도'].isnull().sum()

0

In [105]:
# 성공한 데이터 프레임 저장
df_save2.to_csv(PATH + "/03_금연구역/금연구역_장병용_전처리_성공(37344).csv", index=False, encoding='cp949')

### 무엇이 문제일까 - 실패한 주소지 탐색

In [106]:
df_fail3.head(10)

구            구분      시설명                     주소  \
1   중구      학교교과교습학원  두드림음악학원          계백로1571번길 46    
2   중구  어린이운송용 승합자동차   문화어린이집  대전 보문로 207번길14 문화어린이집   
3   중구  어린이운송용 승합자동차   은혜어린이집    대전 태평동 삼부@ 410동 12호   
4   중구  어린이운송용 승합자동차  꿈나무어린이집           대전 선화로57번길 8   
6   중구  어린이운송용 승합자동차    산성검도관                 보문로55    
7   중구  어린이운송용 승합자동차    한성유도관               대종로 2015   
13  중구        게임제공업소  JUNE PC              목중로54번길 2   
14  중구        게임제공업소    지존PC방              목중로54번길 2   
15  중구        게임제공업소   장수게임랜드            대종로584번길 55   
16  중구        게임제공업소     토이박스                계룡로 785   

                                  번지주소         세부주소                  kakao 주소  \
1     대전광역시 중구 계백로1571번길 46 , 2층 (유천동)          2층           중구 계백로1571번길 46    
2   대전 중구 보문로 207번길14   문화어린이집,  문화유치원        문화유치원  중구 대전 보문로 207번길14 문화어린이집   
3       대전 중구 태평동 삼부@ 410동 12호, 은혜어린이집       은혜어린이집    중구 대전 태평동 삼부@ 410동 12호   
4     대전 중구 선화로57번길 8, 4층(선화동) 꿈나무어린이집   4층 꿈나무어린이집           중구 대전 선화로57번길 8   
6                대전시 중구 보문로55 ,(산성검도관)                              중구 보문로55    
7      대전시 중구 대종로 2015, 한라빌딩 7층(한성유도관)     한라빌딩 7층                중구 대종로 2015   
13      대전광역시 중구 목중로54번길 2, 지하1층 (중촌동)        지하1층               중구 목중로54번길 2   
14        대전광역시 중구 목중로54번길 2, 2층 (중촌동)          2층               중구 목중로54번길 2   
15      대전광역시 중구 대종로584번길 55, 2층 (선화동)          2층             중구 대종로584번길 55   
16       대전광역시 중구 계룡로 785, 1층 1호 (용두동)       1층 1호                 중구 계룡로 785   

      동   위도   경도  
1   NaN  NaN  NaN  
2   NaN  NaN  NaN  
3   NaN  NaN  NaN  
4   NaN  NaN  NaN  
6   NaN  NaN  NaN  
7   NaN  NaN  NaN  
13  NaN  NaN  NaN  
14  NaN  NaN  NaN  
15  NaN  NaN  NaN  
16  NaN  NaN  NaN

#### kakao 주소에서 대전 삭제

In [107]:
# kakao 주소에서 대전 검색

df_fail3[df_fail3['kakao 주소'].str.contains('대전')]

구               구분              시설명                            주소  \
2       중구     어린이운송용 승합자동차           문화어린이집         대전 보문로 207번길14 문화어린이집   
3       중구     어린이운송용 승합자동차           은혜어린이집           대전 태평동 삼부@ 410동 12호   
4       중구     어린이운송용 승합자동차          꿈나무어린이집                  대전 선화로57번길 8   
27      중구  음식점(휴게, 일반, 제과)  우정집실내포장마차앤신존떡볶이                    대전 천서로 583   
87      동구              음식점         미자쌀롱우송대점                   동대전 로 192-1   
...    ...              ...              ...                           ...   
2959   유성구              음식점          청춘아 달리자          도룡동 3번지 8호 대전 무역전시관    
3280   유성구              음식점            보람있는집          방현동 0번지 북대전 IC 졸음쉼터    
7094   유성구           담배소매업소             조은마트  노은동 510번지 7호 대전 회관주차장빌딩.라은빌딩   
7120   유성구           담배소매업소              신혜자     덕명동 산16번지 1호 대전 산업대 잡화코너    
10223  유성구          어린이놀이시설    럭키하나아파트 제3놀이터      가정로 63 대전 가정로 63 럭키하나아파트   

                                                    번지주소         세부주소  \
2                     대전 중구 보문로 207번길14   문화어린이집,  문화유치원        문화유치원   
3                         대전 중구 태평동 삼부@ 410동 12호, 은혜어린이집       은혜어린이집   
4                       대전 중구 선화로57번길 8, 4층(선화동) 꿈나무어린이집   4층 꿈나무어린이집   
27                          대전광역시 중구 대전천서로 583, 1층 (선화동)          1층    
87                         대전광역시 동구 동대전로 192-1, 1층 (자양동)          1층    
...                                                  ...          ...   
2959                       대전광역시 유성구 도룡동 3번지 8호 대전무역전시관           NaN   
3280                    대전광역시 유성구 방현동 0번지 북대전IC(논산)졸음쉼터           NaN   
7094               대전광역시 유성구 노은동 510번지 7호 대전회관주차장빌딩.라은빌딩          NaN   
7120            대전광역시 유성구 덕명동 산16번지 1호 대전산업대 잡화코너 (3층)            NaN   
10223  대전광역시 유성구 가정로 63 (신성동, 럭키하나아파트)대전 유성구 가정로 63(신...          NaN   

                               kakao 주소    동   위도   경도  
2              중구 대전 보문로 207번길14 문화어린이집  NaN  NaN  NaN  
3                중구 대전 태평동 삼부@ 410동 12호  NaN  NaN  NaN  
4                       중구 대전 선화로57번길 8  NaN  NaN  NaN  
27                        중구 대전 천서로 583  NaN  NaN  NaN  
87                       동구 동대전 로 192-1  NaN  NaN  NaN  
...                                 ...  ...  ...  ...  
2959           유성구 도룡동 3번지 8호 대전 무역전시관   NaN  NaN  NaN  
3280           유성구 방현동 0번지 북대전 IC 졸음쉼터   NaN  NaN  NaN  
7094   유성구 노은동 510번지 7호 대전 회관주차장빌딩.라은빌딩  NaN  NaN  NaN  
7120      유성구 덕명동 산16번지 1호 대전 산업대 잡화코너   NaN  NaN  NaN  
10223      유성구 가정로 63 대전 가정로 63 럭키하나아파트  NaN  NaN  NaN  

[74 rows x 10 columns]

In [108]:
# kakao 주소에서 대전 삭제
df_fail3["kakao 주소"] = df_fail3["kakao 주소"].str.replace(pat='대전', repl=r' ', regex=True)
df_fail3[df_fail3['kakao 주소'].str.contains('대전')]

C:\Users\Helpe\AppData\Local\Temp\ipykernel_11692\104256011.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_fail3["kakao 주소"] = df_fail3["kakao 주소"].str.replace(pat='대전', repl=r' ', regex=True)


Empty DataFrame
Columns: [구, 구분, 시설명, 주소, 번지주소, 세부주소, kakao 주소, 동, 위도, 경도]
Index: []

#### 대전 제거 후 카카오 주소 공백제거

In [110]:
print(df_fail3['kakao 주소'].str.contains('  ').sum()) # 스페이스바 두번
print(df_fail3['kakao 주소'].str.contains('   ').sum()) # 스페이스바 세번
print(df_fail3['kakao 주소'].str.contains('    ').sum()) # 스페이스바 네번
print(df_fail3['kakao 주소'].str.contains('     ').sum()) # 스페이스바 다섯번
print(df_fail3['kakao 주소'].str.contains('      ').sum()) # 스페이스바 섯번

74
58
2
2
0


In [114]:
# 공백 제거
df_fail3["kakao 주소"] = df_fail3["kakao 주소"].str.replace(pat='  ', repl=r' ', regex=True)
print(df_fail3['kakao 주소'].str.contains('  ').sum()) # 스페이스바 두번
print(df_fail3['kakao 주소'].str.contains('   ').sum()) # 스페이스바 세번
print(df_fail3['kakao 주소'].str.contains('    ').sum()) # 스페이스바 네번
print(df_fail3['kakao 주소'].str.contains('     ').sum()) # 스페이스바 다섯번
print(df_fail3['kakao 주소'].str.contains('      ').sum()) # 스페이스바 섯번

0
0
0
0
0


In [115]:
df_fail3.reset_index(drop=True, inplace=True)
df_fail3.head()

0         중구 계백로1571번길 46 
1    중구 보문로 207번길14 문화어린이집
2      중구 태평동 삼부@ 410동 12호
3             중구 선화로57번길 8
4                중구 보문로55 
Name: kakao 주소, dtype: object

In [ ]:
df_road = df_fail3.copy()

In [117]:
df_road

: 

: 

In [262]:
road_list = df_road['kakao 주소']
road_list[1]

'중구 계백로1571번길 46 '

In [266]:
kakao.search_address(road_list[4])

{'documents': [],
 'meta': {'is_end': True, 'pageable_count': 0, 'total_count': 0}}

#### 번지주소 체크

* 동 뒤에 공백 삽입
* 중복 공백 제거

In [282]:
df_road[df_road['주소'].str.contains('동')]

구               구분                 시설명                   주소  \
3       중구     어린이운송용 승합자동차              은혜어린이집  대전 태평동 삼부@ 410동 12호   
9       중구     어린이운송용 승합자동차              꾸러기태권도            동부로 55-58   
10      중구     어린이운송용 승합자동차              꿈내리유치원          대전 산성동6-43    
17      중구  음식점(휴게, 일반, 제과)              맥주마을치킨        동서대로1466번길 25   
20      중구  음식점(휴게, 일반, 제과)           멕시카나치킨목동점           목동로22번길 37   
...    ...              ...                 ...                  ...   
11110  유성구             복합용도             나라아이넷사옥             탑립동860-0   
11111  유성구             복합용도  파크팰리스(Park Palace)             탑립동862-0   
11112  유성구             복합용도                  농협             하기동531-1   
11113  유성구             복합용도               우성프라자             하기동531-2   
11114  유성구             복합용도                  식당             학하동676-1   

                                            번지주소          세부주소  \
3                 대전 중구 태평동 삼부@ 410동 12호, 은혜어린이집        은혜어린이집   
9      대전광역시 동구 동부로 55-58, 601동 1401호(판암동 주공아파트)   601동 1401호    
10                대전 중구 산성동6-43(우성아파트 내), 꿈내리유치원        꿈내리유치원   
17              대전광역시 중구 동서대로1466번길 25, 1층 (선화동)           1층    
20                  대전광역시 중구 목동로22번길 37, 1층 (목동)           1층    
...                                          ...           ...   
11110                         대전광역시 유성구 탑립동860-0           NaN   
11111                         대전광역시 유성구 탑립동862-0           NaN   
11112                         대전광역시 유성구 하기동531-1           NaN   
11113                         대전광역시 유성구 하기동531-2           NaN   
11114                         대전광역시 유성구 학하동676-1           NaN   

                  kakao 주소    동   위도   경도  
3      중구 태평동 삼부@ 410동 12호  NaN  NaN  NaN  
9             중구 동부로 55-58  NaN  NaN  NaN  
10             중구 산성동6-43   NaN  NaN  NaN  
17        중구 동서대로1466번길 25  NaN  NaN  NaN  
20           중구 목동로22번길 37  NaN  NaN  NaN  
...                    ...  ...  ...  ...  
11110         유성구 탑립동860-0  NaN  NaN  NaN  
11111         유성구 탑립동862-0  NaN  NaN  NaN  
11112         유성구 하기동531-1  NaN  NaN  NaN  
11113         유성구 하기동531-2  NaN  NaN  NaN  
11114         유성구 학하동676-1  NaN  NaN  NaN  

[9015 rows x 10 columns]

In [283]:
# 동 뒤에 공백 삽입
df_road['주소'] = df_road['주소'].str.replace('동', '동 ')

In [284]:
print(df_road['kakao 주소'].str.contains('  ').sum()) # 스페이스바 두번
print(df_road['kakao 주소'].str.contains('   ').sum()) # 스페이스바 세번

0
0


In [285]:
# 중복 공백 제거
df_road["kakao 주소"] = df_road["kakao 주소"].str.replace(pat='  ', repl=r' ', regex=True)
print(df_road['kakao 주소'].str.contains('  ').sum()) # 스페이스바 두번

0


In [286]:
df_dong = df_road[df_road['kakao 주소'].str.contains('동')].copy()
df_dong.reset_index(drop=True, inplace=True)
df_dong

구               구분                 시설명                     주소  \
0      중구     어린이운송용 승합자동차              은혜어린이집  대전 태평동  삼부@ 410동  12호   
1      중구     어린이운송용 승합자동차              꾸러기태권도             동 부로 55-58   
2      중구     어린이운송용 승합자동차              꿈내리유치원           대전 산성동 6-43    
3      중구  음식점(휴게, 일반, 제과)              맥주마을치킨         동 서대로1466번길 25   
4      중구  음식점(휴게, 일반, 제과)           멕시카나치킨목동점            목동 로22번길 37   
...   ...              ...                 ...                    ...   
9068  유성구             복합용도             나라아이넷사옥              탑립동 860-0   
9069  유성구             복합용도  파크팰리스(Park Palace)              탑립동 862-0   
9070  유성구             복합용도                  농협              하기동 531-1   
9071  유성구             복합용도               우성프라자              하기동 531-2   
9072  유성구             복합용도                  식당              학하동 676-1   

                                           번지주소          세부주소  \
0                대전 중구 태평동 삼부@ 410동 12호, 은혜어린이집        은혜어린이집   
1     대전광역시 동구 동부로 55-58, 601동 1401호(판암동 주공아파트)   601동 1401호    
2                대전 중구 산성동6-43(우성아파트 내), 꿈내리유치원        꿈내리유치원   
3              대전광역시 중구 동서대로1466번길 25, 1층 (선화동)           1층    
4                  대전광역시 중구 목동로22번길 37, 1층 (목동)           1층    
...                                         ...           ...   
9068                         대전광역시 유성구 탑립동860-0           NaN   
9069                         대전광역시 유성구 탑립동862-0           NaN   
9070                         대전광역시 유성구 하기동531-1           NaN   
9071                         대전광역시 유성구 하기동531-2           NaN   
9072                         대전광역시 유성구 학하동676-1           NaN   

                 kakao 주소    동   위도   경도  
0     중구 태평동 삼부@ 410동 12호  NaN  NaN  NaN  
1            중구 동부로 55-58  NaN  NaN  NaN  
2             중구 산성동6-43   NaN  NaN  NaN  
3        중구 동서대로1466번길 25  NaN  NaN  NaN  
4           중구 목동로22번길 37  NaN  NaN  NaN  
...                   ...  ...  ...  ...  
9068         유성구 탑립동860-0  NaN  NaN  NaN  
9069         유성구 탑립동862-0  NaN  NaN  NaN  
9070         유성구 하기동531-1  NaN  NaN  NaN  
9071         유성구 하기동531-2  NaN  NaN  NaN  
9072         유성구 학하동676-1  NaN  NaN  NaN  

[9073 rows x 10 columns]

동부로, 중부로 등 도로명 주소 사이에 공백이 삽입된 케이스 존재

In [290]:
df_dong['주소'].str[:5].unique()

array(['대전 태평', '동 부로 ', '대전 산성', '동 서대로', '목동 로2', '동 성대로', '동 춘당로',
       '흥룡로 4', '만년동  ', '가수원동 ', '둔산동  ', '대전 청로', '고니길 4', '천동 초등',
       '효천4길 ', '동 대전 ', '효천1길 ', '우암로31', '새둑길 3', '충정로13', '계족로43',
       '대전 천동', '새둑길 7', '갑동  3', '갑동  4', '계산동  ', '관평동  ', '구암동  ',
       '궁동  4', '궁동  2', '노은동  ', '대정동  ', '덕명동  ', '덕진동  ', '도룡동  ',
       '문지동  ', '반석동  ', '방현동  ', '복용동  ', '봉명동  ', '상대동  ', '송강동  ',
       '신성동  ', '용계동  ', '원내동  ', '원신흥동 ', '원촌동  ', '자운동  ', '장대동  ',
       '장동  3', '전민동  ', '죽동  7', '죽동  6', '지족동  ', '추목동  ', '탑립동  ',
       '하기동  ', '학하동  ', '대전 목동', '문화동  ', '대전 부사', '대전 유천', '대전 문화',
       '대전 옥계', '대전 호동', '대전 중촌', '대전 계백', '대전 동 ', '산성동  ', '목동 로4',
       '유천동 3', '유천동 로', '대전 용두', '대전 도솔', '용두동 2', '용두동  ', '유등천동 ',
       '선화동  ', '부사동  ', '유천동  ', '문창동  ', '보문산로3', '대전 송촌', '신탄진동 ',
       '대전 오정', '대전 석봉', '송촌동 2', '법동 28', '법동  3', '상서동  ', '중리동  ',
       '장동  2', '오정동  ', '목상동  ', '읍내동  ', '법동  1', '송촌동  ', '관저동  ',
       '도마동  ', '용문동

In [293]:

df_dong['주소'] = df_dong['주소'].str.replace('동 부', '동부')
df_dong['주소'] = df_dong['주소'].str.replace('동 춘', '동춘')
df_dong['주소'] = df_dong['주소'].str.replace('동 서', '동서')
df_dong['주소'] = df_dong['주소'].str.replace('천동 초', '천동초')
df_dong['주소'] = df_dong['주소'].str.replace('동 성', '동성')
df_dong['주소'] = df_dong['주소'].str.replace('동 대', '동대')
df_dong['주소'] = df_dong['주소'].str.replace('동 로', '동로')

In [295]:
df_dong['주소'].str[:4].unique()

array(['대전 태', '동부로 ', '대전 산', '동서대로', '목동로2', '동성대로', '동춘당로', '흥룡로 ',
       '만년동 ', '가수원동', '둔산동 ', '대전 청', '고니길 ', '천동초등', '효천4길', '동대전 ',
       '효천1길', '우암로3', '새둑길 ', '충정로1', '계족로4', '대전 천', '갑동  ', '계산동 ',
       '관평동 ', '구암동 ', '궁동  ', '노은동 ', '대정동 ', '덕명동 ', '덕진동 ', '도룡동 ',
       '문지동 ', '반석동 ', '방현동 ', '복용동 ', '봉명동 ', '상대동 ', '송강동 ', '신성동 ',
       '용계동 ', '원내동 ', '원신흥동', '원촌동 ', '자운동 ', '장대동 ', '장동  ', '전민동 ',
       '죽동  ', '지족동 ', '추목동 ', '탑립동 ', '하기동 ', '학하동 ', '대전 목', '문화동 ',
       '대전 부', '대전 유', '대전 문', '대전 옥', '대전 호', '대전 중', '대전 계', '대전 동',
       '산성동 ', '목동로4', '유천동 ', '유천동로', '대전 용', '대전 도', '용두동 ', '유등천동',
       '선화동 ', '부사동 ', '문창동 ', '보문산로', '대전 송', '신탄진동', '대전 오', '대전 석',
       '송촌동 ', '법동 2', '법동  ', '상서동 ', '중리동 ', '오정동 ', '목상동 ', '읍내동 ',
       '관저동 ', '도마동 ', '용문동 ', '갈마동 ', '탄방동 ', '괴정동 ', '내동  ', '월평동 ',
       '변동  ', '복수동 ', '정림동 ', '흑석동 ', '가장동 ', '갈마2동', '계백로 ', '관저동로',
       '복수동로', '도안동로', '월평동로', '평촌동 ', '도마1동', '괴곡동 ', '도마2동', '둔산1동',
      

In [296]:
df_dong

구               구분                 시설명                     주소  \
0      중구     어린이운송용 승합자동차              은혜어린이집  대전 태평동  삼부@ 410동  12호   
1      중구     어린이운송용 승합자동차              꾸러기태권도              동부로 55-58   
2      중구     어린이운송용 승합자동차              꿈내리유치원           대전 산성동 6-43    
3      중구  음식점(휴게, 일반, 제과)              맥주마을치킨          동서대로1466번길 25   
4      중구  음식점(휴게, 일반, 제과)           멕시카나치킨목동점             목동로22번길 37   
...   ...              ...                 ...                    ...   
9068  유성구             복합용도             나라아이넷사옥              탑립동 860-0   
9069  유성구             복합용도  파크팰리스(Park Palace)              탑립동 862-0   
9070  유성구             복합용도                  농협              하기동 531-1   
9071  유성구             복합용도               우성프라자              하기동 531-2   
9072  유성구             복합용도                  식당              학하동 676-1   

                                           번지주소          세부주소  \
0                대전 중구 태평동 삼부@ 410동 12호, 은혜어린이집        은혜어린이집   
1     대전광역시 동구 동부로 55-58, 601동 1401호(판암동 주공아파트)   601동 1401호    
2                대전 중구 산성동6-43(우성아파트 내), 꿈내리유치원        꿈내리유치원   
3              대전광역시 중구 동서대로1466번길 25, 1층 (선화동)           1층    
4                  대전광역시 중구 목동로22번길 37, 1층 (목동)           1층    
...                                         ...           ...   
9068                         대전광역시 유성구 탑립동860-0           NaN   
9069                         대전광역시 유성구 탑립동862-0           NaN   
9070                         대전광역시 유성구 하기동531-1           NaN   
9071                         대전광역시 유성구 하기동531-2           NaN   
9072                         대전광역시 유성구 학하동676-1           NaN   

                 kakao 주소    동   위도   경도  
0     중구 태평동 삼부@ 410동 12호  NaN  NaN  NaN  
1            중구 동부로 55-58  NaN  NaN  NaN  
2             중구 산성동6-43   NaN  NaN  NaN  
3        중구 동서대로1466번길 25  NaN  NaN  NaN  
4           중구 목동로22번길 37  NaN  NaN  NaN  
...                   ...  ...  ...  ...  
9068         유성구 탑립동860-0  NaN  NaN  NaN  
9069         유성구 탑립동862-0  NaN  NaN  NaN  
9070         유성구 하기동531-1  NaN  NaN  NaN  
9071         유성구 하기동531-2  NaN  NaN  NaN  
9072         유성구 학하동676-1  NaN  NaN  NaN  

[9073 rows x 10 columns]

#### 구 + 주소 결합

In [298]:
df_dong['kakao 주소'] = df_dong['구'].str.cat(df_dong['주소'], sep=' ')
df_dong

구               구분                 시설명                     주소  \
0      중구     어린이운송용 승합자동차              은혜어린이집  대전 태평동  삼부@ 410동  12호   
1      중구     어린이운송용 승합자동차              꾸러기태권도              동부로 55-58   
2      중구     어린이운송용 승합자동차              꿈내리유치원           대전 산성동 6-43    
3      중구  음식점(휴게, 일반, 제과)              맥주마을치킨          동서대로1466번길 25   
4      중구  음식점(휴게, 일반, 제과)           멕시카나치킨목동점             목동로22번길 37   
...   ...              ...                 ...                    ...   
9068  유성구             복합용도             나라아이넷사옥              탑립동 860-0   
9069  유성구             복합용도  파크팰리스(Park Palace)              탑립동 862-0   
9070  유성구             복합용도                  농협              하기동 531-1   
9071  유성구             복합용도               우성프라자              하기동 531-2   
9072  유성구             복합용도                  식당              학하동 676-1   

                                           번지주소          세부주소  \
0                대전 중구 태평동 삼부@ 410동 12호, 은혜어린이집        은혜어린이집   
1     대전광역시 동구 동부로 55-58, 601동 1401호(판암동 주공아파트)   601동 1401호    
2                대전 중구 산성동6-43(우성아파트 내), 꿈내리유치원        꿈내리유치원   
3              대전광역시 중구 동서대로1466번길 25, 1층 (선화동)           1층    
4                  대전광역시 중구 목동로22번길 37, 1층 (목동)           1층    
...                                         ...           ...   
9068                         대전광역시 유성구 탑립동860-0           NaN   
9069                         대전광역시 유성구 탑립동862-0           NaN   
9070                         대전광역시 유성구 하기동531-1           NaN   
9071                         대전광역시 유성구 하기동531-2           NaN   
9072                         대전광역시 유성구 학하동676-1           NaN   

                      kakao 주소    동   위도   경도  
0     중구 대전 태평동  삼부@ 410동  12호  NaN  NaN  NaN  
1                 중구 동부로 55-58  NaN  NaN  NaN  
2              중구 대전 산성동 6-43   NaN  NaN  NaN  
3             중구 동서대로1466번길 25  NaN  NaN  NaN  
4                중구 목동로22번길 37  NaN  NaN  NaN  
...                        ...  ...  ...  ...  
9068             유성구 탑립동 860-0  NaN  NaN  NaN  
9069             유성구 탑립동 862-0  NaN  NaN  NaN  
9070             유성구 하기동 531-1  NaN  NaN  NaN  
9071             유성구 하기동 531-2  NaN  NaN  NaN  
9072             유성구 학하동 676-1  NaN  NaN  NaN  

[9073 rows x 10 columns]

In [299]:
# 대전 제거
df_dong["kakao 주소"] = df_dong["kakao 주소"].str.replace(pat='대전', repl=r' ', regex=True)
df_dong[df_dong['kakao 주소'].str.contains('대전')]

Empty DataFrame
Columns: [구, 구분, 시설명, 주소, 번지주소, 세부주소, kakao 주소, 동, 위도, 경도]
Index: []

In [303]:
road_list = df_dong['kakao 주소']
kakao.search_address(road_list[9072])

{'documents': [{'address': {'address_name': '대전 유성구 학하동 676-1',
    'b_code': '3020010500',
    'h_code': '3020052600',
    'main_address_no': '676',
    'mountain_yn': 'N',
    'region_1depth_name': '대전',
    'region_2depth_name': '유성구',
    'region_3depth_h_name': '학하동',
    'region_3depth_name': '학하동',
    'sub_address_no': '1',
    'x': '127.305220060679',
    'y': '36.3328417231939'},
   'address_name': '대전 유성구 학하동 676-1',
   'address_type': 'REGION_ADDR',
   'road_address': {'address_name': '대전 유성구 진잠옛로135번길 115',
    'building_name': '사회복지법인신생원',
    'main_building_no': '115',
    'region_1depth_name': '대전',
    'region_2depth_name': '유성구',
    'region_3depth_name': '학하동',
    'road_name': '진잠옛로135번길',
    'sub_building_no': '',
    'underground_yn': 'N',
    'x': '127.304575520238',
    'y': '36.3332445726416',
    'zone_no': '34215'},
   'x': '127.305220060679',
   'y': '36.3328417231939'}],
 'meta': {'is_end': True, 'pageable_count': 1, 'total_count': 1}}

지번의 경우 인식에 성공한듯?  
다시 api 호출 세팅


In [304]:
for i in tqdm(range(len(road_list))):
    try :
        df_dong['동'].iloc[i] = kakao.search_address(road_list[i])['documents'][0]['address']['region_3depth_name']
        df_dong['경도'].iloc[i] = kakao.search_address(road_list[i])['documents'][0]['x']
        df_dong['위도'].iloc[i] = kakao.search_address(road_list[i])['documents'][0]['y']
    except:
        pass


100%|██████████| 9073/9073 [13:16<00:00, 11.40it/s]  


In [305]:
df_dong.isnull().sum()

구              0
구분             0
시설명            0
주소             0
번지주소           0
세부주소        8646
kakao 주소       0
동           6288
위도          6288
경도          6288
dtype: int64

2785개 추가로 위경도 추출에 성공함

#### 성공한 데이터프레임 저장

In [332]:
df_fail2 = df_dong[df_dong['위도'].isnull()]
df_save_dong = df_dong[~df_dong['위도'].isnull()]

print(df_fail2.shape)
print(df_save_dong.shape)

(6288, 10)
(2785, 10)


In [318]:
# 병합
df_save2= pd.concat([df_save_dong, df_save], 
                    axis=0,
                    join='outer'
                    )


df_save2.reset_index(drop=True)

구      구분             시설명                                 주소  \
0       동구  사회복지시설  은어송마을1단지(아)경로당                           대전 청로 67   
1       동구  사회복지시설       미리내(아)경로당                      대전 청로203번길 35   
2       동구  사회복지시설  은어송마을2단지(아)경로당                           대전 청로 35   
3       동구  사회복지시설        해드림실버요양원                           대전 청로 95   
4      유성구     음식점           탐라돌구이                 봉명동  607번지 5호 지상1층   
...    ...     ...             ...                                ...   
35532  유성구     음식점             순카페    관평동 666번지 테크노벨리4단지상가 101호 102호    
35533  유성구     음식점      선비꼬마김밥 관평점  관평동 668번지 대덕테크노밸리3단지아파트상가동 101-1호   
35534  유성구     음식점            김밥도둑   관평동 669번지 대덕테크노밸리5단지 분산상가동 104호    
35535  유성구     음식점       가마치통닭 관평점         관평동 669번지 분산상가동 지상1층 101호    
35536  유성구     음식점          삼촌네불고기   관평동 672번지 대덕테크노밸리6단지아파트 상가동 106호   

                                               번지주소         세부주소  \
0                      대전 동구 동구청로 67, 은어송마을1단지(가오동)    은어송마을1단지    
1                      대전 동구 동구청로203번길 35,미리내@(판암동)        미리내@    
2                     대전 동구 동구청로 35, 은어송마을2단지@(대성동)   은어송마을2단지@    
3                           대전 동구 동구청로 95, 8층 (가오동)          8층    
4            대전광역시 유성구 봉명동 607번지 5호 지상1층, 101,102호     101,102호    
...                                             ...          ...   
35532     대전광역시 유성구 관평동 666번지 테크노벨리4단지상가 101호 102호           NaN   
35533   대전광역시 유성구 관평동 668번지 대덕테크노밸리3단지아파트상가동 101-1호          NaN   
35534  대전광역시 유성구 관평동 669번지 대덕테크노밸리5단지(아)분산상가동 104호           NaN   
35535          대전광역시 유성구 관평동 669번지 분산상가동 지상1층 101호           NaN   
35536    대전광역시 유성구 관평동 672번지 대덕테크노밸리6단지아파트 상가동 106호          NaN   

                                    kakao 주소    동                위도  \
0                                 동구   청로 67  가오동  36.3056962273781   
1                            동구   청로203번길 35  판암동  36.3159127178655   
2                                 동구   청로 35  대성동  36.3037469941274   
3                                 동구   청로 95  가오동  36.3065069481971   
4                     유성구 봉명동  607번지 5호 지상1층   호동  36.3025866691176   
...                                      ...  ...               ...   
35532    유성구 관평동 666번지 테크노벨리4단지상가 101호 102호   관평동  36.4195098553081   
35533  유성구 관평동 668번지 대덕테크노밸리3단지아파트상가동 101-1호  관평동  36.4212223376349   
35534   유성구 관평동 669번지 대덕테크노밸리5단지 분산상가동 104호   관평동  36.4200754766059   
35535         유성구 관평동 669번지 분산상가동 지상1층 101호   관평동  36.4200754766059   
35536   유성구 관평동 672번지 대덕테크노밸리6단지아파트 상가동 106호  관평동  36.4193494654472   

                     경도  
0      127.457215836363  
1      127.455110655522  
2      127.458270508282  
3      127.456739020314  
4      127.447484466761  
...                 ...  
35532   127.38451778774  
35533  127.385635590115  
35534  127.387262769268  
35535  127.387262769268  
35536  127.390124429871  

[35537 rows x 10 columns]

In [309]:
df_save2.shape

(35537, 10)

#### 2차 csv 저장

In [311]:
df_save2.to_csv(PATH + "/03_금연구역/금연구역_장병용_전처리_성공(35537).csv", index=False, encoding='cp949')
df_fail2.to_csv(PATH + "/03_금연구역/금연구역_장병용_전처리_실패(6288).csv", index=False, encoding='cp949')

In [314]:
df_fail2.groupby('구분')['주소'].count().sort_values(ascending=False) 


구분
음식점                  4322
담배판매업소 및 담배자동판매기      560
담배소매업소                537
의료기관                  315
일반음식점 및 휴게음식점         250
연면적 1천제곱미터 이상 건축물      83
어린이운송용 승합자동차           37
신고체육시설                 33
어린이놀이시설                31
목욕장                    19
어린이집                   18
담배소매인업소                10
복합건축물                  10
음식점(휴게, 일반, 제과)         9
학원                      9
복합용도                    9
사회복지시설                  7
사무용건축물                  7
어린이운송용승합차               6
게임제공업소                  4
청사                      3
실내체육시설                  2
청소년활동시설                 2
체육시설                    2
어린이운송용차량                1
도서관                     1
학교                      1
Name: 주소, dtype: int64